{
 "cells": [
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "Hua Sun\n",
    "\n",
    "PDX Pan-Cancer manuscript\n",
    "\n",
    "\n",
    "v2 2020-05-19;6/10; 8/23\n",
    "\n",
    "Alteration of druggable genes that associated with gene expression \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##======================== 0. Set library"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "── Attaching packages ─────────────────────────────────────── tidyverse 1.2.1 ──\n",
      "✔ ggplot2 3.0.0     ✔ purrr   0.2.5\n",
      "✔ tibble  1.4.2     ✔ dplyr   0.7.6\n",
      "✔ tidyr   0.8.1     ✔ stringr 1.3.1\n",
      "✔ readr   1.1.1     ✔ forcats 0.3.0\n",
      "── Conflicts ────────────────────────────────────────── tidyverse_conflicts() ──\n",
      "✖ dplyr::between()   masks data.table::between()\n",
      "✖ dplyr::filter()    masks stats::filter()\n",
      "✖ dplyr::first()     masks data.table::first()\n",
      "✖ dplyr::lag()       masks stats::lag()\n",
      "✖ dplyr::last()      masks data.table::last()\n",
      "✖ purrr::transpose() masks data.table::transpose()\n",
      "------------------------------------------------------------------------------\n",
      "You have loaded plyr after dplyr - this is likely to cause problems.\n",
      "If you need functions from both plyr and dplyr, please load plyr first, then dplyr:\n",
      "library(plyr); library(dplyr)\n",
      "------------------------------------------------------------------------------\n",
      "\n",
      "Attaching package: ‘plyr’\n",
      "\n",
      "The following objects are masked from ‘package:dplyr’:\n",
      "\n",
      "    arrange, count, desc, failwith, id, mutate, rename, summarise,\n",
      "    summarize\n",
      "\n",
      "The following object is masked from ‘package:purrr’:\n",
      "\n",
      "    compact\n",
      "\n",
      "\n",
      "Attaching package: ‘reshape2’\n",
      "\n",
      "The following object is masked from ‘package:tidyr’:\n",
      "\n",
      "    smiths\n",
      "\n",
      "The following objects are masked from ‘package:data.table’:\n",
      "\n",
      "    dcast, melt\n",
      "\n",
      "Loading required package: magrittr\n",
      "\n",
      "Attaching package: ‘magrittr’\n",
      "\n",
      "The following object is masked from ‘package:purrr’:\n",
      "\n",
      "    set_names\n",
      "\n",
      "The following object is masked from ‘package:tidyr’:\n",
      "\n",
      "    extract\n",
      "\n",
      "\n",
      "Attaching package: ‘ggpubr’\n",
      "\n",
      "The following object is masked from ‘package:plyr’:\n",
      "\n",
      "    mutate\n",
      "\n",
      "\n",
      "********************************************************\n",
      "Note: As of version 1.0.0, cowplot does not change the\n",
      "  default ggplot2 theme anymore. To recover the previous\n",
      "  behavior, execute:\n",
      "  theme_set(theme_cowplot())\n",
      "********************************************************\n",
      "\n",
      "\n",
      "Attaching package: ‘cowplot’\n",
      "\n",
      "The following object is masked from ‘package:ggpubr’:\n",
      "\n",
      "    get_legend\n",
      "\n"
     ]
    }
   ],
   "source": [
    "\n",
    "library(data.table)\n",
    "library(tidyverse)\n",
    "library(tidyr)\n",
    "library(plyr)\n",
    "library(reshape2)\n",
    "library(ggpubr)\n",
    "library(cowplot) # use merged plot\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 0. data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# input files\n",
    "f_drugList <- \"~/Documents/2.proj-PDX/2.analysis_pdxEvolution/analysis-therapeuticEvent/v2/_nci-match/nci-match_eay131.20200821.formatted.table\"\n",
    "f_somaticMut <- \"~/Documents/2.proj-PDX/2.analysis_pdxEvolution/v5.datafreeze/2.somaticMut/datafreeze.somaticMut.all.non-silent.meta3.formal.v5.20200427.tsv\"\n",
    "f_cnv <- \"~/Documents/2.proj-PDX/2.analysis_pdxEvolution/v5.datafreeze/3.cnv/dataFreezeSamples.cnvkit.geneLevel.from_segment.v5.tsv\"\n",
    "f_geneExp <- \"~/Documents/2.proj-PDX/2.analysis_pdxEvolution/v5.datafreeze/5.geneExp/pdmr_other_washU-b1-b8.datafreeze.v5.kallisto.tsv\"\n",
    "f_fusion <- \"~/Documents/2.proj-PDX/2.analysis_pdxEvolution/v5.datafreeze/6.fusion/star-fusion.merged.FusionInspector.fusions.abridged.annotated.coding_effect.filtered.v5.tsv\" \n",
    "#f_sampleInfo <- \"~/Documents/2.proj-PDX/2.analysis_pdxEvolution/v5.datafreeze/1.sampleInfo/sampleInfo.washU_b1-b8.pdmr.other.passed.v5.extra.20200401.v2.tsv\" \n",
    "f_sampleInfo <- \"~/Documents/2.proj-PDX/2.analysis_pdxEvolution/0.sampleInfo/4.merged/v5/sampleInfo.washU_b1-b8.pdmr.other.passed.v5.extra.20200401.v4.1.tsv\" \n",
    "f_info_wxs_matched_rna <- \"~/Documents/2.proj-PDX/2.analysis_pdxEvolution/0.sampleInfo/4.merged/v5/pair_table/all.WES_with_RNASEQ.HT_PDX.sampleinfo.tsv\"\n",
    "\n",
    "# read files\n",
    "d_drugSet <- fread(f_drugList, data.table=F)\n",
    "d_geneExp <- fread(f_geneExp, data.table=F)\n",
    "d_sampleInfo <- fread(f_sampleInfo, data.table=F)\n",
    "d_info_wxs_rna <- fread(f_info_wxs_matched_rna, data.table=F)\n",
    "\n",
    "# make PDX only data_info that WES with RNA-Seq\n",
    "d_pdx_info <- d_info_wxs_rna[d_info_wxs_rna$Group==\"PDX\",]\n",
    "    # Source CancerGroup CancerType CaseID  SampleID Group Passage RNASEQ  WES\n",
    "\n",
    "\n",
    "# output analysis results\n",
    "outdir <- \".\"\n",
    "if (! dir.exists(outdir)) { dir.create(outdir) }\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Set functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "#--------- (Important checking) make text file for record siginficant amplification\n",
    "\n",
    "##======= set functions for checking significantly diff mut in exp\n",
    "\n",
    "func_rec_sigDiffExp_in_alt <- function(df, alteration){\n",
    "    # df\n",
    "    # CancerType, SampleID, gene, altGroup, log2tpm\n",
    "    \n",
    "    d_rec <- setNames(data.frame(matrix(ncol = 7, nrow = 0)), c(\"CancerType\", \"Gene\", \"Alt_event\", \"n_total\", \"n_Ref\", \"n_Alt\", \"wilcoxon.p-val\"))\n",
    "    alt_event <- alteration\n",
    "    \n",
    "    for(cancerType in unique(df$CancerType)){\n",
    "    \n",
    "        df_cancer <- df[df$CancerType==cancerType,]\n",
    "        # filter sample size < 10\n",
    "        if(length(unique(df_cancer$SampleID))<10){ next }\n",
    "        \n",
    "        for(i in unique(df$gene)){\n",
    "            \n",
    "            df2 <- df_cancer[df_cancer$gene==i,]\n",
    "            \n",
    "            # remove genes \n",
    "            # if no any mut then del. (for pdx 1 sample also possible)\n",
    "            if(alteration == 'Fusion'){\n",
    "                n_total <- length(unique(df2$RNASEQ))\n",
    "                n_alt <- length(unique(df2$RNASEQ[df2$altGroup==alteration]))\n",
    "            } else {\n",
    "                n_total <- length(unique(df2$WES))\n",
    "                n_alt <- length(unique(df2$WES[df2$altGroup==alteration]))\n",
    "            }\n",
    "            \n",
    "            if(n_alt < 3){ next }\n",
    "            \n",
    "            # set altGroup must 2\n",
    "            if(length(unique(df2$altGroup))<2){next}\n",
    "            \n",
    "            # 2.not significant in gene exp. between two groups\n",
    "            w_test <- wilcox.test(log2tpm ~ altGroup, data = df2)\n",
    "            if(w_test$p.value >= 0.05){ next }\n",
    "\n",
    "            # output significant\n",
    "            if (w_test$p.value < 0.05){\n",
    "                p_val <- formatC(w_test$p.value, format = \"e\", digits = 1)\n",
    "               # fdr <- p.adjust(p_val, method = \"fdr\", n = length(p_val))\n",
    "                n_ref <- n_total - n_alt\n",
    "                str <- c(cancerType, i, alt_event, n_total, n_ref, n_alt, p_val)\n",
    "                d_rec <- rbindlist(list(d_rec, as.list(str)))\n",
    "            }\n",
    "        }\n",
    "    }\n",
    "\n",
    "    # output\n",
    "    if (is.data.frame(d_rec) && nrow(d_rec)==0) { return('empty')}\n",
    "    d_rec <- d_rec[order(Gene, CancerType),]\n",
    "    \n",
    "    \n",
    "    return(d_rec)\n",
    "}\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 1. Somatic mutations and gene expression"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<thead><tr><th scope=col>gene</th><th scope=col>WES</th><th scope=col>RNASEQ</th><th scope=col>tpm</th><th scope=col>log2tpm</th><th scope=col>Source</th><th scope=col>CancerGroup</th><th scope=col>CancerType</th><th scope=col>CaseID</th><th scope=col>SampleID</th><th scope=col>Group</th><th scope=col>Passage</th><th scope=col>mut</th><th scope=col>PDXmodelID</th><th scope=col>altGroup</th></tr></thead>\n",
       "<tbody>\n",
       "\t<tr><td>AKT1                        </td><td>10960X12_HCI-009_PDXtumor_P3</td><td>10962X12_HCI009_PDXtumor_P3 </td><td>120.23551                   </td><td>6.921669                    </td><td>HCI                         </td><td>Breast                      </td><td>BRCA                        </td><td>HCI-HCI009                  </td><td>HCI-HCI009.P3               </td><td>PDX                         </td><td>P3                          </td><td>NA                          </td><td>HCI009                      </td><td>WT                          </td></tr>\n",
       "\t<tr><td>AKT1                        </td><td>10960X13_HCI-009_PDXtumor_P6</td><td>10962X13_HCI009_PDXtumor_P6 </td><td>110.21338                   </td><td>6.797187                    </td><td>HCI                         </td><td>Breast                      </td><td>BRCA                        </td><td>HCI-HCI009                  </td><td>HCI-HCI009.P6               </td><td>PDX                         </td><td>P6                          </td><td>NA                          </td><td>HCI009                      </td><td>WT                          </td></tr>\n",
       "\t<tr><td>AKT1                        </td><td>10960X14_HCI-010_PDXtumor_P3</td><td>10962X14_HCI010_PDXtumor_P3 </td><td> 33.39358                   </td><td>5.104067                    </td><td>HCI                         </td><td>Breast                      </td><td>BRCA                        </td><td>HCI-HCI010                  </td><td>HCI-HCI010.P3               </td><td>PDX                         </td><td>P3                          </td><td>NA                          </td><td>HCI010                      </td><td>WT                          </td></tr>\n",
       "\t<tr><td>AKT1                        </td><td>10960X15_HCI011_PDXtumor_P3 </td><td>10962X15_HCI011_PDXtumor_P3 </td><td>151.11013                   </td><td>7.248972                    </td><td>HCI                         </td><td>Breast                      </td><td>BRCA                        </td><td>HCI-HCI011                  </td><td>HCI-HCI011.P3               </td><td>PDX                         </td><td>P3                          </td><td>NA                          </td><td>HCI011                      </td><td>WT                          </td></tr>\n",
       "\t<tr><td>AKT1                        </td><td>10960X16_HCI012_PDXtumor_P4 </td><td>10962X16_HCI012_PDXtumor_P4 </td><td>184.41844                   </td><td>7.534641                    </td><td>HCI                         </td><td>Breast                      </td><td>BRCA                        </td><td>HCI-HCI012                  </td><td>HCI-HCI012.P4               </td><td>PDX                         </td><td>P4                          </td><td>NA                          </td><td>HCI012                      </td><td>WT                          </td></tr>\n",
       "\t<tr><td>AKT1                        </td><td>10960X17_HCI013_PDXtumor_P3 </td><td>10962X17_HCI013_PDXtumor_P3 </td><td> 19.79917                   </td><td>4.378454                    </td><td>HCI                         </td><td>Breast                      </td><td>BRCA                        </td><td>HCI-HCI013                  </td><td>HCI-HCI013.P3               </td><td>PDX                         </td><td>P3                          </td><td>NA                          </td><td>HCI013                      </td><td>WT                          </td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "\\begin{tabular}{r|lllllllllllllll}\n",
       " gene & WES & RNASEQ & tpm & log2tpm & Source & CancerGroup & CancerType & CaseID & SampleID & Group & Passage & mut & PDXmodelID & altGroup\\\\\n",
       "\\hline\n",
       "\t AKT1                               & 10960X12\\_HCI-009\\_PDXtumor\\_P3 & 10962X12\\_HCI009\\_PDXtumor\\_P3  & 120.23551                          & 6.921669                           & HCI                                & Breast                             & BRCA                               & HCI-HCI009                         & HCI-HCI009.P3                      & PDX                                & P3                                 & NA                                 & HCI009                             & WT                                \\\\\n",
       "\t AKT1                               & 10960X13\\_HCI-009\\_PDXtumor\\_P6 & 10962X13\\_HCI009\\_PDXtumor\\_P6  & 110.21338                          & 6.797187                           & HCI                                & Breast                             & BRCA                               & HCI-HCI009                         & HCI-HCI009.P6                      & PDX                                & P6                                 & NA                                 & HCI009                             & WT                                \\\\\n",
       "\t AKT1                               & 10960X14\\_HCI-010\\_PDXtumor\\_P3 & 10962X14\\_HCI010\\_PDXtumor\\_P3  &  33.39358                          & 5.104067                           & HCI                                & Breast                             & BRCA                               & HCI-HCI010                         & HCI-HCI010.P3                      & PDX                                & P3                                 & NA                                 & HCI010                             & WT                                \\\\\n",
       "\t AKT1                               & 10960X15\\_HCI011\\_PDXtumor\\_P3  & 10962X15\\_HCI011\\_PDXtumor\\_P3  & 151.11013                          & 7.248972                           & HCI                                & Breast                             & BRCA                               & HCI-HCI011                         & HCI-HCI011.P3                      & PDX                                & P3                                 & NA                                 & HCI011                             & WT                                \\\\\n",
       "\t AKT1                               & 10960X16\\_HCI012\\_PDXtumor\\_P4  & 10962X16\\_HCI012\\_PDXtumor\\_P4  & 184.41844                          & 7.534641                           & HCI                                & Breast                             & BRCA                               & HCI-HCI012                         & HCI-HCI012.P4                      & PDX                                & P4                                 & NA                                 & HCI012                             & WT                                \\\\\n",
       "\t AKT1                               & 10960X17\\_HCI013\\_PDXtumor\\_P3  & 10962X17\\_HCI013\\_PDXtumor\\_P3  &  19.79917                          & 4.378454                           & HCI                                & Breast                             & BRCA                               & HCI-HCI013                         & HCI-HCI013.P3                      & PDX                                & P3                                 & NA                                 & HCI013                             & WT                                \\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "gene | WES | RNASEQ | tpm | log2tpm | Source | CancerGroup | CancerType | CaseID | SampleID | Group | Passage | mut | PDXmodelID | altGroup | \n",
       "|---|---|---|---|---|---|\n",
       "| AKT1                         | 10960X12_HCI-009_PDXtumor_P3 | 10962X12_HCI009_PDXtumor_P3  | 120.23551                    | 6.921669                     | HCI                          | Breast                       | BRCA                         | HCI-HCI009                   | HCI-HCI009.P3                | PDX                          | P3                           | NA                           | HCI009                       | WT                           | \n",
       "| AKT1                         | 10960X13_HCI-009_PDXtumor_P6 | 10962X13_HCI009_PDXtumor_P6  | 110.21338                    | 6.797187                     | HCI                          | Breast                       | BRCA                         | HCI-HCI009                   | HCI-HCI009.P6                | PDX                          | P6                           | NA                           | HCI009                       | WT                           | \n",
       "| AKT1                         | 10960X14_HCI-010_PDXtumor_P3 | 10962X14_HCI010_PDXtumor_P3  |  33.39358                    | 5.104067                     | HCI                          | Breast                       | BRCA                         | HCI-HCI010                   | HCI-HCI010.P3                | PDX                          | P3                           | NA                           | HCI010                       | WT                           | \n",
       "| AKT1                         | 10960X15_HCI011_PDXtumor_P3  | 10962X15_HCI011_PDXtumor_P3  | 151.11013                    | 7.248972                     | HCI                          | Breast                       | BRCA                         | HCI-HCI011                   | HCI-HCI011.P3                | PDX                          | P3                           | NA                           | HCI011                       | WT                           | \n",
       "| AKT1                         | 10960X16_HCI012_PDXtumor_P4  | 10962X16_HCI012_PDXtumor_P4  | 184.41844                    | 7.534641                     | HCI                          | Breast                       | BRCA                         | HCI-HCI012                   | HCI-HCI012.P4                | PDX                          | P4                           | NA                           | HCI012                       | WT                           | \n",
       "| AKT1                         | 10960X17_HCI013_PDXtumor_P3  | 10962X17_HCI013_PDXtumor_P3  |  19.79917                    | 4.378454                     | HCI                          | Breast                       | BRCA                         | HCI-HCI013                   | HCI-HCI013.P3                | PDX                          | P3                           | NA                           | HCI013                       | WT                           | \n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "  gene WES                          RNASEQ                      tpm      \n",
       "1 AKT1 10960X12_HCI-009_PDXtumor_P3 10962X12_HCI009_PDXtumor_P3 120.23551\n",
       "2 AKT1 10960X13_HCI-009_PDXtumor_P6 10962X13_HCI009_PDXtumor_P6 110.21338\n",
       "3 AKT1 10960X14_HCI-010_PDXtumor_P3 10962X14_HCI010_PDXtumor_P3  33.39358\n",
       "4 AKT1 10960X15_HCI011_PDXtumor_P3  10962X15_HCI011_PDXtumor_P3 151.11013\n",
       "5 AKT1 10960X16_HCI012_PDXtumor_P4  10962X16_HCI012_PDXtumor_P4 184.41844\n",
       "6 AKT1 10960X17_HCI013_PDXtumor_P3  10962X17_HCI013_PDXtumor_P3  19.79917\n",
       "  log2tpm  Source CancerGroup CancerType CaseID     SampleID      Group Passage\n",
       "1 6.921669 HCI    Breast      BRCA       HCI-HCI009 HCI-HCI009.P3 PDX   P3     \n",
       "2 6.797187 HCI    Breast      BRCA       HCI-HCI009 HCI-HCI009.P6 PDX   P6     \n",
       "3 5.104067 HCI    Breast      BRCA       HCI-HCI010 HCI-HCI010.P3 PDX   P3     \n",
       "4 7.248972 HCI    Breast      BRCA       HCI-HCI011 HCI-HCI011.P3 PDX   P3     \n",
       "5 7.534641 HCI    Breast      BRCA       HCI-HCI012 HCI-HCI012.P4 PDX   P4     \n",
       "6 4.378454 HCI    Breast      BRCA       HCI-HCI013 HCI-HCI013.P3 PDX   P3     \n",
       "  mut PDXmodelID altGroup\n",
       "1 NA  HCI009     WT      \n",
       "2 NA  HCI009     WT      \n",
       "3 NA  HCI010     WT      \n",
       "4 NA  HCI011     WT      \n",
       "5 NA  HCI012     WT      \n",
       "6 NA  HCI013     WT      "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "\n",
    "# outdir\n",
    "outdir_sub <- \"../1.somaticMut_with_geneExp\"\n",
    "if (! dir.exists(outdir_sub)) { dir.create(outdir_sub) }\n",
    "\n",
    "\n",
    "#---------- a) extract drug target\n",
    "mut_tarGeneSet <- d_drugSet[d_drugSet$Event==\"Mutation\",]$Gene\n",
    "\n",
    "\n",
    "#---------- b) extract mutation\n",
    "d_somaticMut <- fread(f_somaticMut, data.table=F)\n",
    "\n",
    "d_somaticMut_tar <- d_somaticMut %>%\n",
    "                        filter(Tumor_Sample_Barcode %in% d_pdx_info$WES) %>%\n",
    "                        filter(Hugo_Symbol %in% mut_tarGeneSet)\n",
    "\n",
    "d_somaticMut_tar <- d_somaticMut_tar[,c('Tumor_Sample_Barcode', 'Hugo_Symbol', 'HGVSp_Short')]\n",
    "rownames(d_somaticMut_tar) <- NULL\n",
    "\n",
    "# merge to one gene if the gene has multiple mutations\n",
    "# gene mut1,mut2,...\n",
    "d_somaticMut_tar <- ddply(d_somaticMut_tar, .(Tumor_Sample_Barcode, Hugo_Symbol), summarize,\n",
    "                        HGVSp_Short=paste(HGVSp_Short, collapse=\",\")\n",
    "                    )\n",
    "colnames(d_somaticMut_tar) <- c('WES', 'gene', 'mut')\n",
    "# WES gene mut\n",
    "\n",
    "\n",
    "# output text file for checking\n",
    "write.table(d_somaticMut_tar, paste0(outdir_sub,'/1.all.target.somaticMut.',Sys.Date(),'.tsv'), sep = \"\\t\", row.names=FALSE, quote = FALSE)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "#---------- c) extract gene expression\n",
    "\n",
    "# format gene exp data frame as --> sample gene log2tpm\n",
    "d_geneExp_tar <- subset(d_geneExp, Name %in% mut_tarGeneSet)\n",
    "d_geneExp_tar <- reshape2::melt(d_geneExp_tar, id.vars=c(\"Name\"))\n",
    "colnames(d_geneExp_tar) <- c('gene', 'RNASEQ', 'tpm')\n",
    "d_geneExp_tar <- d_geneExp_tar[, c('RNASEQ', 'gene', 'tpm')]\n",
    "# RNASEQ gene tpm\n",
    "\n",
    "# tpm --> log2(tpm+1)\n",
    "d_geneExp_tar$log2tpm <- log2(d_geneExp_tar$tpm + 1)\n",
    "\n",
    "d_geneExp_tar <- d_geneExp_tar %>% filter(RNASEQ %in% d_pdx_info$RNASEQ)\n",
    "rownames(d_geneExp_tar) <- NULL\n",
    "# RNASEQ gene tpm log2tpm\n",
    "\n",
    "\n",
    "#---------- d) merge formatted gene exp with somatic mut datafrme\n",
    "\n",
    "#NOTE- must mut data merge to gene exp rather than gene exp merge to mut\n",
    "\n",
    "#1. merge data info\n",
    "d_tar_expMut <- merge(d_geneExp_tar, d_pdx_info, on='RNASEQ', all.x=TRUE)\n",
    "#2. merge mut\n",
    "d_tar_expMut <- merge(d_tar_expMut, d_somaticMut_tar, on=c('WES', 'gene'), all.x=TRUE)\n",
    "\n",
    "#3. add PDXmodelID\n",
    "d_tar_expMut$PDXmodelID <- d_sampleInfo$ModelID_proper[match(d_tar_expMut$WES, d_sampleInfo$Analysis_ID)]\n",
    "\n",
    "\n",
    "\n",
    "#---------- e) add flag with Mut & Non-mut\n",
    "d_tar_expMut$altGroup <- 'WT'\n",
    "d_tar_expMut$altGroup[d_tar_expMut$mut != 'NA'] <- 'Mutation'\n",
    "\n",
    "\n",
    "# output text file for checking\n",
    "write.table(d_tar_expMut, paste0(outdir_sub,'/2.all.target.geneExp_with_somaticMut.',Sys.Date(),'.tsv'), sep = \"\\t\", row.names=FALSE, quote = FALSE)\n",
    "\n",
    "head(d_tar_expMut)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "#---------- f) make visulization (1.set plot function)\n",
    "\n",
    "library(ggplot2)\n",
    "library(ggpubr)\n",
    "\n",
    "\n",
    "## two group with statistic\n",
    "CompBoxplot2exp <- function(title, dataFrame){\n",
    "    \n",
    "    #max.y <- max(dataFrame$log2tpm)\n",
    "    #n_mut_case <- length(unique(dataFrame$CaseID[dataFrame$altGroup=='Mutation']))\n",
    "    #n_other_case <- length(unique(dataFrame$CaseID[dataFrame$altGroup=='WT']))\n",
    "    n_mut_case <- length(unique(dataFrame$PDXmodelID[dataFrame$altGroup=='Mutation']))\n",
    "    n_other_case <- length(unique(dataFrame$PDXmodelID[dataFrame$altGroup=='WT']))\n",
    "    \n",
    "    n_mut_sample <- length(unique(dataFrame$WES[dataFrame$altGroup=='Mutation']))\n",
    "    n_other_sample <- length(unique(dataFrame$WES[dataFrame$altGroup=='WT']))\n",
    "    \n",
    "    label_mut <- paste0(\"Mutation\",\"\\nN=\",n_mut_case,\"\\nn=\",n_mut_sample)\n",
    "    label_other <- paste0(\"WT\",\"\\nN=\",n_other_case,\"\\nn=\",n_other_sample)\n",
    "    \n",
    "\n",
    "    \n",
    "#    n_mut <- nrow(dataFrame[dataFrame$altGroup=='Mutation',])\n",
    "#    n_other <- nrow(dataFrame[dataFrame$altGroup=='Other',])\n",
    "    \n",
    "    p <- ggplot(dataFrame, aes(x=altGroup, y=log2tpm, color=altGroup)) +\n",
    "            geom_boxplot(lwd=0.6, outlier.size=0.6) + \n",
    "            geom_jitter(position=position_jitter(0.1), size=0.6, alpha = 0.5) +\n",
    "            theme_classic() +\n",
    "            scale_color_manual(values = c(\"WT\" = \"#B5B5B5\", \"Mutation\" = \"#4EAF49\")) + \n",
    "            theme(legend.position=\"None\") +\n",
    "            ggtitle(title) +\n",
    "            xlab(\"\") +\n",
    "            ylab(\"Gene expression\\nlog2(TPM+1)\") + \n",
    "            scale_x_discrete(\n",
    "                limit = c(\"WT\", \"Mutation\"),\n",
    "                labels=c(\"WT\" = label_other, \"Mutation\" = label_mut)\n",
    "            ) +\n",
    "            theme(\n",
    "                plot.title = element_text(size=7, hjust = 0.5),\n",
    "                axis.text = element_text(size = 6),\n",
    "                axis.title=element_text(size=6)) + \n",
    "            theme(\n",
    "                axis.ticks = element_line(size = rel(0.5)),\n",
    "                axis.ticks.length=unit(1.5, \"mm\"),\n",
    "                axis.line = element_line(colour = 'black', size = 0.2)\n",
    "            ) +\n",
    "            theme(panel.grid.major = element_blank(), \n",
    "               panel.grid.minor = element_blank(), \n",
    "               panel.background = element_blank(), \n",
    "               panel.border = element_blank()\n",
    "            )\n",
    "    \n",
    "    #p2 <- p + stat_compare_means(size = 1, method = \"wilcox.test\", label.x = 1, label.y = max.y*0.9)\n",
    "    \n",
    "    \n",
    "    \n",
    "    return(p)      \n",
    "}\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "#---------- f) make visulization (2.plotting)\n",
    "\n",
    "\n",
    "for(cancerType in unique(d_tar_expMut$CancerType)){\n",
    "    \n",
    "    df_cancer <- d_tar_expMut[d_tar_expMut$CancerType==cancerType,]\n",
    "    # filter sample size < 10\n",
    "    if(length(unique(df_cancer$SampleID))<10){ next }\n",
    "    \n",
    "    \n",
    "    pdf(paste0(outdir_sub,'/',cancerType,'.mut_geneExp.',Sys.Date(),'.pdf'), width = 1.4, height = 1.8, useDingbats=FALSE)\n",
    "\n",
    "    for(i in unique(d_tar_expMut$gene)){\n",
    "    \n",
    "        df2 <- df_cancer[df_cancer$gene==i,]\n",
    "        mutationSite <- unique(df2$mut[df2$altGroup!=\"WT\"])\n",
    "        #print(i)\n",
    "    \n",
    "        # remove genes \n",
    "        # if no any mut then del. (for pdx 1 sample also possible)\n",
    "        if(nrow(df2[df2$altGroup=='Mutation',])==0){ \n",
    "            #print(paste0(\"[Warning] No mutation in the gene \", i, \" !\"))\n",
    "            next\n",
    "        }\n",
    "        \n",
    "        # set altGroup must 2\n",
    "        if(length(unique(df2$altGroup))<2){next}\n",
    "        \n",
    "        # 2.not significant in gene exp. between two groups\n",
    "        w_test <- wilcox.test(log2tpm ~ altGroup, data = df2)\n",
    "        if(w_test$p.value >= 0.05){\n",
    "            #print(paste0(\"[Warning] Not significant in the gene \", i, \" \", w_test$p.value))\n",
    "            next\n",
    "        }\n",
    "    \n",
    "    \n",
    "        # output significant\n",
    "        if (nrow(df2[df2$altGroup=='Mutation',])>0 && w_test$p.value < 0.05){\n",
    "            \n",
    "            p_val <- formatC(w_test$p.value, format = \"e\", digits = 1)\n",
    "           # q_val <- p.adjust(p_val, method = \"fdr\", n = length(p_val))\n",
    "            title <- paste0(i,':',mutationSite , ' (', cancerType, ')', \"\\nWilcoxon, p=\", p_val)\n",
    "            \n",
    "            #title <- paste0(i, ' (', cancerType, ')')\n",
    "            #p_val <- paste0('Wilcoxon, p=', format(w_test$p.value, scientific = TRUE))\n",
    "            p1 <- CompBoxplot2exp(title, df2)\n",
    "            #p_passage <- CompBoxplot2exp_passage(title, df2)\n",
    "    \n",
    "            # merged plot\n",
    "            #p_merged <- plot_grid(p1, p_passage, nrow=1, align=\"h\", rel_widths = c(1, 2), labels = c('A', 'B'))\n",
    "    \n",
    "            #print(p_merged)\n",
    "            print(p1)\n",
    "        }\n",
    "    \n",
    "    }\n",
    "\n",
    "    dev.off()\n",
    "\n",
    "}\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<thead><tr><th scope=col>gene</th><th scope=col>WES</th><th scope=col>RNASEQ</th><th scope=col>tpm</th><th scope=col>log2tpm</th><th scope=col>Source</th><th scope=col>CancerGroup</th><th scope=col>CancerType</th><th scope=col>CaseID</th><th scope=col>SampleID</th><th scope=col>Group</th><th scope=col>Passage</th><th scope=col>mut</th><th scope=col>PDXmodelID</th><th scope=col>altGroup</th></tr></thead>\n",
       "<tbody>\n",
       "\t<tr><td>AKT1                        </td><td>10960X12_HCI-009_PDXtumor_P3</td><td>10962X12_HCI009_PDXtumor_P3 </td><td>120.23551                   </td><td>6.921669                    </td><td>HCI                         </td><td>Breast                      </td><td>BRCA                        </td><td>HCI-HCI009                  </td><td>HCI-HCI009.P3               </td><td>PDX                         </td><td>P3                          </td><td>NA                          </td><td>HCI009                      </td><td>WT                          </td></tr>\n",
       "\t<tr><td>AKT1                        </td><td>10960X13_HCI-009_PDXtumor_P6</td><td>10962X13_HCI009_PDXtumor_P6 </td><td>110.21338                   </td><td>6.797187                    </td><td>HCI                         </td><td>Breast                      </td><td>BRCA                        </td><td>HCI-HCI009                  </td><td>HCI-HCI009.P6               </td><td>PDX                         </td><td>P6                          </td><td>NA                          </td><td>HCI009                      </td><td>WT                          </td></tr>\n",
       "\t<tr><td>AKT1                        </td><td>10960X14_HCI-010_PDXtumor_P3</td><td>10962X14_HCI010_PDXtumor_P3 </td><td> 33.39358                   </td><td>5.104067                    </td><td>HCI                         </td><td>Breast                      </td><td>BRCA                        </td><td>HCI-HCI010                  </td><td>HCI-HCI010.P3               </td><td>PDX                         </td><td>P3                          </td><td>NA                          </td><td>HCI010                      </td><td>WT                          </td></tr>\n",
       "\t<tr><td>AKT1                        </td><td>10960X15_HCI011_PDXtumor_P3 </td><td>10962X15_HCI011_PDXtumor_P3 </td><td>151.11013                   </td><td>7.248972                    </td><td>HCI                         </td><td>Breast                      </td><td>BRCA                        </td><td>HCI-HCI011                  </td><td>HCI-HCI011.P3               </td><td>PDX                         </td><td>P3                          </td><td>NA                          </td><td>HCI011                      </td><td>WT                          </td></tr>\n",
       "\t<tr><td>AKT1                        </td><td>10960X16_HCI012_PDXtumor_P4 </td><td>10962X16_HCI012_PDXtumor_P4 </td><td>184.41844                   </td><td>7.534641                    </td><td>HCI                         </td><td>Breast                      </td><td>BRCA                        </td><td>HCI-HCI012                  </td><td>HCI-HCI012.P4               </td><td>PDX                         </td><td>P4                          </td><td>NA                          </td><td>HCI012                      </td><td>WT                          </td></tr>\n",
       "\t<tr><td>AKT1                        </td><td>10960X17_HCI013_PDXtumor_P3 </td><td>10962X17_HCI013_PDXtumor_P3 </td><td> 19.79917                   </td><td>4.378454                    </td><td>HCI                         </td><td>Breast                      </td><td>BRCA                        </td><td>HCI-HCI013                  </td><td>HCI-HCI013.P3               </td><td>PDX                         </td><td>P3                          </td><td>NA                          </td><td>HCI013                      </td><td>WT                          </td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "\\begin{tabular}{r|lllllllllllllll}\n",
       " gene & WES & RNASEQ & tpm & log2tpm & Source & CancerGroup & CancerType & CaseID & SampleID & Group & Passage & mut & PDXmodelID & altGroup\\\\\n",
       "\\hline\n",
       "\t AKT1                               & 10960X12\\_HCI-009\\_PDXtumor\\_P3 & 10962X12\\_HCI009\\_PDXtumor\\_P3  & 120.23551                          & 6.921669                           & HCI                                & Breast                             & BRCA                               & HCI-HCI009                         & HCI-HCI009.P3                      & PDX                                & P3                                 & NA                                 & HCI009                             & WT                                \\\\\n",
       "\t AKT1                               & 10960X13\\_HCI-009\\_PDXtumor\\_P6 & 10962X13\\_HCI009\\_PDXtumor\\_P6  & 110.21338                          & 6.797187                           & HCI                                & Breast                             & BRCA                               & HCI-HCI009                         & HCI-HCI009.P6                      & PDX                                & P6                                 & NA                                 & HCI009                             & WT                                \\\\\n",
       "\t AKT1                               & 10960X14\\_HCI-010\\_PDXtumor\\_P3 & 10962X14\\_HCI010\\_PDXtumor\\_P3  &  33.39358                          & 5.104067                           & HCI                                & Breast                             & BRCA                               & HCI-HCI010                         & HCI-HCI010.P3                      & PDX                                & P3                                 & NA                                 & HCI010                             & WT                                \\\\\n",
       "\t AKT1                               & 10960X15\\_HCI011\\_PDXtumor\\_P3  & 10962X15\\_HCI011\\_PDXtumor\\_P3  & 151.11013                          & 7.248972                           & HCI                                & Breast                             & BRCA                               & HCI-HCI011                         & HCI-HCI011.P3                      & PDX                                & P3                                 & NA                                 & HCI011                             & WT                                \\\\\n",
       "\t AKT1                               & 10960X16\\_HCI012\\_PDXtumor\\_P4  & 10962X16\\_HCI012\\_PDXtumor\\_P4  & 184.41844                          & 7.534641                           & HCI                                & Breast                             & BRCA                               & HCI-HCI012                         & HCI-HCI012.P4                      & PDX                                & P4                                 & NA                                 & HCI012                             & WT                                \\\\\n",
       "\t AKT1                               & 10960X17\\_HCI013\\_PDXtumor\\_P3  & 10962X17\\_HCI013\\_PDXtumor\\_P3  &  19.79917                          & 4.378454                           & HCI                                & Breast                             & BRCA                               & HCI-HCI013                         & HCI-HCI013.P3                      & PDX                                & P3                                 & NA                                 & HCI013                             & WT                                \\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "gene | WES | RNASEQ | tpm | log2tpm | Source | CancerGroup | CancerType | CaseID | SampleID | Group | Passage | mut | PDXmodelID | altGroup | \n",
       "|---|---|---|---|---|---|\n",
       "| AKT1                         | 10960X12_HCI-009_PDXtumor_P3 | 10962X12_HCI009_PDXtumor_P3  | 120.23551                    | 6.921669                     | HCI                          | Breast                       | BRCA                         | HCI-HCI009                   | HCI-HCI009.P3                | PDX                          | P3                           | NA                           | HCI009                       | WT                           | \n",
       "| AKT1                         | 10960X13_HCI-009_PDXtumor_P6 | 10962X13_HCI009_PDXtumor_P6  | 110.21338                    | 6.797187                     | HCI                          | Breast                       | BRCA                         | HCI-HCI009                   | HCI-HCI009.P6                | PDX                          | P6                           | NA                           | HCI009                       | WT                           | \n",
       "| AKT1                         | 10960X14_HCI-010_PDXtumor_P3 | 10962X14_HCI010_PDXtumor_P3  |  33.39358                    | 5.104067                     | HCI                          | Breast                       | BRCA                         | HCI-HCI010                   | HCI-HCI010.P3                | PDX                          | P3                           | NA                           | HCI010                       | WT                           | \n",
       "| AKT1                         | 10960X15_HCI011_PDXtumor_P3  | 10962X15_HCI011_PDXtumor_P3  | 151.11013                    | 7.248972                     | HCI                          | Breast                       | BRCA                         | HCI-HCI011                   | HCI-HCI011.P3                | PDX                          | P3                           | NA                           | HCI011                       | WT                           | \n",
       "| AKT1                         | 10960X16_HCI012_PDXtumor_P4  | 10962X16_HCI012_PDXtumor_P4  | 184.41844                    | 7.534641                     | HCI                          | Breast                       | BRCA                         | HCI-HCI012                   | HCI-HCI012.P4                | PDX                          | P4                           | NA                           | HCI012                       | WT                           | \n",
       "| AKT1                         | 10960X17_HCI013_PDXtumor_P3  | 10962X17_HCI013_PDXtumor_P3  |  19.79917                    | 4.378454                     | HCI                          | Breast                       | BRCA                         | HCI-HCI013                   | HCI-HCI013.P3                | PDX                          | P3                           | NA                           | HCI013                       | WT                           | \n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "  gene WES                          RNASEQ                      tpm      \n",
       "1 AKT1 10960X12_HCI-009_PDXtumor_P3 10962X12_HCI009_PDXtumor_P3 120.23551\n",
       "2 AKT1 10960X13_HCI-009_PDXtumor_P6 10962X13_HCI009_PDXtumor_P6 110.21338\n",
       "3 AKT1 10960X14_HCI-010_PDXtumor_P3 10962X14_HCI010_PDXtumor_P3  33.39358\n",
       "4 AKT1 10960X15_HCI011_PDXtumor_P3  10962X15_HCI011_PDXtumor_P3 151.11013\n",
       "5 AKT1 10960X16_HCI012_PDXtumor_P4  10962X16_HCI012_PDXtumor_P4 184.41844\n",
       "6 AKT1 10960X17_HCI013_PDXtumor_P3  10962X17_HCI013_PDXtumor_P3  19.79917\n",
       "  log2tpm  Source CancerGroup CancerType CaseID     SampleID      Group Passage\n",
       "1 6.921669 HCI    Breast      BRCA       HCI-HCI009 HCI-HCI009.P3 PDX   P3     \n",
       "2 6.797187 HCI    Breast      BRCA       HCI-HCI009 HCI-HCI009.P6 PDX   P6     \n",
       "3 5.104067 HCI    Breast      BRCA       HCI-HCI010 HCI-HCI010.P3 PDX   P3     \n",
       "4 7.248972 HCI    Breast      BRCA       HCI-HCI011 HCI-HCI011.P3 PDX   P3     \n",
       "5 7.534641 HCI    Breast      BRCA       HCI-HCI012 HCI-HCI012.P4 PDX   P4     \n",
       "6 4.378454 HCI    Breast      BRCA       HCI-HCI013 HCI-HCI013.P3 PDX   P3     \n",
       "  mut PDXmodelID altGroup\n",
       "1 NA  HCI009     WT      \n",
       "2 NA  HCI009     WT      \n",
       "3 NA  HCI010     WT      \n",
       "4 NA  HCI011     WT      \n",
       "5 NA  HCI012     WT      \n",
       "6 NA  HCI013     WT      "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#--------- (Important checking) make text file for record siginficant amplification\n",
    "alterationEvent <- 'Mutation'\n",
    "\n",
    "head(d_tar_expMut)\n",
    "\n",
    "d_rec <- func_rec_sigDiffExp_in_alt(d_tar_expMut, alterationEvent)\n",
    "\n",
    "write.table(d_rec, paste0(outdir_sub,'/1.p-val.target.geneExp_with_',alterationEvent,'.',Sys.Date(),'.tsv'), sep = \"\\t\", row.names=FALSE, quote = FALSE)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 2. Fusions and gene expression"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<thead><tr><th scope=col>RNASEQ</th><th scope=col>gene</th><th scope=col>fusionName</th><th scope=col>fusionType</th></tr></thead>\n",
       "<tbody>\n",
       "\t<tr><td>114868_125_R_FCWC50_v1_2_RNASEQ      </td><td>FGFR3                                </td><td>FGFR3--TACC3                         </td><td>INFRAME                              </td></tr>\n",
       "\t<tr><td>114868_125_R_FCWC69_v1_2_RNASEQ      </td><td>FGFR3                                </td><td>FGFR3--TACC3                         </td><td>INFRAME                              </td></tr>\n",
       "\t<tr><td>114868_125_R_FEWC21_v1_2_RNASEQ      </td><td>FGFR3                                </td><td>FGFR3--TACC3                         </td><td>INFRAME                              </td></tr>\n",
       "\t<tr><td>114868_125_R_FEWC31_v1_2_RNASEQ      </td><td>FGFR3                                </td><td>FGFR3--TACC3                         </td><td>INFRAME                              </td></tr>\n",
       "\t<tr><td>182917_245_R_JWP_v2_0_1_4_0_RNASEQ   </td><td>BRAF                                 </td><td>BRAF--POT1,BCAP29--BRAF              </td><td>INFRAME,INFRAME                      </td></tr>\n",
       "\t<tr><td>182917_245_R_JXKE22_v2_0_1_4_0_RNASEQ</td><td>BRAF                                 </td><td>BCAP29--BRAF,BRAF--POT1              </td><td>INFRAME,INFRAME                      </td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "\\begin{tabular}{r|llll}\n",
       " RNASEQ & gene & fusionName & fusionType\\\\\n",
       "\\hline\n",
       "\t 114868\\_125\\_R\\_FCWC50\\_v1\\_2\\_RNASEQ       & FGFR3                                             & FGFR3--TACC3                                      & INFRAME                                          \\\\\n",
       "\t 114868\\_125\\_R\\_FCWC69\\_v1\\_2\\_RNASEQ       & FGFR3                                             & FGFR3--TACC3                                      & INFRAME                                          \\\\\n",
       "\t 114868\\_125\\_R\\_FEWC21\\_v1\\_2\\_RNASEQ       & FGFR3                                             & FGFR3--TACC3                                      & INFRAME                                          \\\\\n",
       "\t 114868\\_125\\_R\\_FEWC31\\_v1\\_2\\_RNASEQ       & FGFR3                                             & FGFR3--TACC3                                      & INFRAME                                          \\\\\n",
       "\t 182917\\_245\\_R\\_JWP\\_v2\\_0\\_1\\_4\\_0\\_RNASEQ    & BRAF                                                    & BRAF--POT1,BCAP29--BRAF                                 & INFRAME,INFRAME                                        \\\\\n",
       "\t 182917\\_245\\_R\\_JXKE22\\_v2\\_0\\_1\\_4\\_0\\_RNASEQ & BRAF                                                    & BCAP29--BRAF,BRAF--POT1                                 & INFRAME,INFRAME                                        \\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "RNASEQ | gene | fusionName | fusionType | \n",
       "|---|---|---|---|---|---|\n",
       "| 114868_125_R_FCWC50_v1_2_RNASEQ       | FGFR3                                 | FGFR3--TACC3                          | INFRAME                               | \n",
       "| 114868_125_R_FCWC69_v1_2_RNASEQ       | FGFR3                                 | FGFR3--TACC3                          | INFRAME                               | \n",
       "| 114868_125_R_FEWC21_v1_2_RNASEQ       | FGFR3                                 | FGFR3--TACC3                          | INFRAME                               | \n",
       "| 114868_125_R_FEWC31_v1_2_RNASEQ       | FGFR3                                 | FGFR3--TACC3                          | INFRAME                               | \n",
       "| 182917_245_R_JWP_v2_0_1_4_0_RNASEQ    | BRAF                                  | BRAF--POT1,BCAP29--BRAF               | INFRAME,INFRAME                       | \n",
       "| 182917_245_R_JXKE22_v2_0_1_4_0_RNASEQ | BRAF                                  | BCAP29--BRAF,BRAF--POT1               | INFRAME,INFRAME                       | \n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "  RNASEQ                                gene  fusionName             \n",
       "1 114868_125_R_FCWC50_v1_2_RNASEQ       FGFR3 FGFR3--TACC3           \n",
       "2 114868_125_R_FCWC69_v1_2_RNASEQ       FGFR3 FGFR3--TACC3           \n",
       "3 114868_125_R_FEWC21_v1_2_RNASEQ       FGFR3 FGFR3--TACC3           \n",
       "4 114868_125_R_FEWC31_v1_2_RNASEQ       FGFR3 FGFR3--TACC3           \n",
       "5 182917_245_R_JWP_v2_0_1_4_0_RNASEQ    BRAF  BRAF--POT1,BCAP29--BRAF\n",
       "6 182917_245_R_JXKE22_v2_0_1_4_0_RNASEQ BRAF  BCAP29--BRAF,BRAF--POT1\n",
       "  fusionType     \n",
       "1 INFRAME        \n",
       "2 INFRAME        \n",
       "3 INFRAME        \n",
       "4 INFRAME        \n",
       "5 INFRAME,INFRAME\n",
       "6 INFRAME,INFRAME"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<thead><tr><th scope=col>RNASEQ</th><th scope=col>gene</th><th scope=col>tpm</th><th scope=col>log2tpm</th><th scope=col>Source</th><th scope=col>CancerGroup</th><th scope=col>CancerType</th><th scope=col>Formal_CaseID</th><th scope=col>NewSampleID</th><th scope=col>Group</th><th scope=col>NCI_Passage</th><th scope=col>ModelID_proper_v2</th><th scope=col>fusionName</th><th scope=col>fusionType</th><th scope=col>altGroup</th></tr></thead>\n",
       "<tbody>\n",
       "\t<tr><td>10962X12_HCI009_PDXtumor_P3</td><td>ALK                        </td><td>  0.000000                 </td><td>0.000000                   </td><td>HCI                        </td><td>Breast                     </td><td>BRCA                       </td><td>HCI-HCI009                 </td><td>HCI-HCI009.P3              </td><td>PDX                        </td><td>P3                         </td><td>HCI009                     </td><td>NA                         </td><td>NA                         </td><td>WT                         </td></tr>\n",
       "\t<tr><td>10962X12_HCI009_PDXtumor_P3</td><td>BRAF                       </td><td>  9.491537                 </td><td>3.391154                   </td><td>HCI                        </td><td>Breast                     </td><td>BRCA                       </td><td>HCI-HCI009                 </td><td>HCI-HCI009.P3              </td><td>PDX                        </td><td>P3                         </td><td>HCI009                     </td><td>NA                         </td><td>NA                         </td><td>WT                         </td></tr>\n",
       "\t<tr><td>10962X12_HCI009_PDXtumor_P3</td><td>FGFR1                      </td><td>  7.551949                 </td><td>3.096253                   </td><td>HCI                        </td><td>Breast                     </td><td>BRCA                       </td><td>HCI-HCI009                 </td><td>HCI-HCI009.P3              </td><td>PDX                        </td><td>P3                         </td><td>HCI009                     </td><td>NA                         </td><td>NA                         </td><td>WT                         </td></tr>\n",
       "\t<tr><td>10962X12_HCI009_PDXtumor_P3</td><td>FGFR2                      </td><td>  1.076378                 </td><td>1.054069                   </td><td>HCI                        </td><td>Breast                     </td><td>BRCA                       </td><td>HCI-HCI009                 </td><td>HCI-HCI009.P3              </td><td>PDX                        </td><td>P3                         </td><td>HCI009                     </td><td>NA                         </td><td>NA                         </td><td>WT                         </td></tr>\n",
       "\t<tr><td>10962X12_HCI009_PDXtumor_P3</td><td>FGFR3                      </td><td> 26.904380                 </td><td>4.802420                   </td><td>HCI                        </td><td>Breast                     </td><td>BRCA                       </td><td>HCI-HCI009                 </td><td>HCI-HCI009.P3              </td><td>PDX                        </td><td>P3                         </td><td>HCI009                     </td><td>NA                         </td><td>NA                         </td><td>WT                         </td></tr>\n",
       "\t<tr><td>10962X12_HCI009_PDXtumor_P3</td><td>FGFR4                      </td><td>126.784226                 </td><td>6.997566                   </td><td>HCI                        </td><td>Breast                     </td><td>BRCA                       </td><td>HCI-HCI009                 </td><td>HCI-HCI009.P3              </td><td>PDX                        </td><td>P3                         </td><td>HCI009                     </td><td>NA                         </td><td>NA                         </td><td>WT                         </td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "\\begin{tabular}{r|lllllllllllllll}\n",
       " RNASEQ & gene & tpm & log2tpm & Source & CancerGroup & CancerType & Formal\\_CaseID & NewSampleID & Group & NCI\\_Passage & ModelID\\_proper\\_v2 & fusionName & fusionType & altGroup\\\\\n",
       "\\hline\n",
       "\t 10962X12\\_HCI009\\_PDXtumor\\_P3 & ALK                               &   0.000000                        & 0.000000                          & HCI                               & Breast                            & BRCA                              & HCI-HCI009                        & HCI-HCI009.P3                     & PDX                               & P3                                & HCI009                            & NA                                & NA                                & WT                               \\\\\n",
       "\t 10962X12\\_HCI009\\_PDXtumor\\_P3 & BRAF                              &   9.491537                        & 3.391154                          & HCI                               & Breast                            & BRCA                              & HCI-HCI009                        & HCI-HCI009.P3                     & PDX                               & P3                                & HCI009                            & NA                                & NA                                & WT                               \\\\\n",
       "\t 10962X12\\_HCI009\\_PDXtumor\\_P3 & FGFR1                             &   7.551949                        & 3.096253                          & HCI                               & Breast                            & BRCA                              & HCI-HCI009                        & HCI-HCI009.P3                     & PDX                               & P3                                & HCI009                            & NA                                & NA                                & WT                               \\\\\n",
       "\t 10962X12\\_HCI009\\_PDXtumor\\_P3 & FGFR2                             &   1.076378                        & 1.054069                          & HCI                               & Breast                            & BRCA                              & HCI-HCI009                        & HCI-HCI009.P3                     & PDX                               & P3                                & HCI009                            & NA                                & NA                                & WT                               \\\\\n",
       "\t 10962X12\\_HCI009\\_PDXtumor\\_P3 & FGFR3                             &  26.904380                        & 4.802420                          & HCI                               & Breast                            & BRCA                              & HCI-HCI009                        & HCI-HCI009.P3                     & PDX                               & P3                                & HCI009                            & NA                                & NA                                & WT                               \\\\\n",
       "\t 10962X12\\_HCI009\\_PDXtumor\\_P3 & FGFR4                             & 126.784226                        & 6.997566                          & HCI                               & Breast                            & BRCA                              & HCI-HCI009                        & HCI-HCI009.P3                     & PDX                               & P3                                & HCI009                            & NA                                & NA                                & WT                               \\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "RNASEQ | gene | tpm | log2tpm | Source | CancerGroup | CancerType | Formal_CaseID | NewSampleID | Group | NCI_Passage | ModelID_proper_v2 | fusionName | fusionType | altGroup | \n",
       "|---|---|---|---|---|---|\n",
       "| 10962X12_HCI009_PDXtumor_P3 | ALK                         |   0.000000                  | 0.000000                    | HCI                         | Breast                      | BRCA                        | HCI-HCI009                  | HCI-HCI009.P3               | PDX                         | P3                          | HCI009                      | NA                          | NA                          | WT                          | \n",
       "| 10962X12_HCI009_PDXtumor_P3 | BRAF                        |   9.491537                  | 3.391154                    | HCI                         | Breast                      | BRCA                        | HCI-HCI009                  | HCI-HCI009.P3               | PDX                         | P3                          | HCI009                      | NA                          | NA                          | WT                          | \n",
       "| 10962X12_HCI009_PDXtumor_P3 | FGFR1                       |   7.551949                  | 3.096253                    | HCI                         | Breast                      | BRCA                        | HCI-HCI009                  | HCI-HCI009.P3               | PDX                         | P3                          | HCI009                      | NA                          | NA                          | WT                          | \n",
       "| 10962X12_HCI009_PDXtumor_P3 | FGFR2                       |   1.076378                  | 1.054069                    | HCI                         | Breast                      | BRCA                        | HCI-HCI009                  | HCI-HCI009.P3               | PDX                         | P3                          | HCI009                      | NA                          | NA                          | WT                          | \n",
       "| 10962X12_HCI009_PDXtumor_P3 | FGFR3                       |  26.904380                  | 4.802420                    | HCI                         | Breast                      | BRCA                        | HCI-HCI009                  | HCI-HCI009.P3               | PDX                         | P3                          | HCI009                      | NA                          | NA                          | WT                          | \n",
       "| 10962X12_HCI009_PDXtumor_P3 | FGFR4                       | 126.784226                  | 6.997566                    | HCI                         | Breast                      | BRCA                        | HCI-HCI009                  | HCI-HCI009.P3               | PDX                         | P3                          | HCI009                      | NA                          | NA                          | WT                          | \n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "  RNASEQ                      gene  tpm        log2tpm  Source CancerGroup\n",
       "1 10962X12_HCI009_PDXtumor_P3 ALK     0.000000 0.000000 HCI    Breast     \n",
       "2 10962X12_HCI009_PDXtumor_P3 BRAF    9.491537 3.391154 HCI    Breast     \n",
       "3 10962X12_HCI009_PDXtumor_P3 FGFR1   7.551949 3.096253 HCI    Breast     \n",
       "4 10962X12_HCI009_PDXtumor_P3 FGFR2   1.076378 1.054069 HCI    Breast     \n",
       "5 10962X12_HCI009_PDXtumor_P3 FGFR3  26.904380 4.802420 HCI    Breast     \n",
       "6 10962X12_HCI009_PDXtumor_P3 FGFR4 126.784226 6.997566 HCI    Breast     \n",
       "  CancerType Formal_CaseID NewSampleID   Group NCI_Passage ModelID_proper_v2\n",
       "1 BRCA       HCI-HCI009    HCI-HCI009.P3 PDX   P3          HCI009           \n",
       "2 BRCA       HCI-HCI009    HCI-HCI009.P3 PDX   P3          HCI009           \n",
       "3 BRCA       HCI-HCI009    HCI-HCI009.P3 PDX   P3          HCI009           \n",
       "4 BRCA       HCI-HCI009    HCI-HCI009.P3 PDX   P3          HCI009           \n",
       "5 BRCA       HCI-HCI009    HCI-HCI009.P3 PDX   P3          HCI009           \n",
       "6 BRCA       HCI-HCI009    HCI-HCI009.P3 PDX   P3          HCI009           \n",
       "  fusionName fusionType altGroup\n",
       "1 NA         NA         WT      \n",
       "2 NA         NA         WT      \n",
       "3 NA         NA         WT      \n",
       "4 NA         NA         WT      \n",
       "5 NA         NA         WT      \n",
       "6 NA         NA         WT      "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                                      RNASEQ  gene       tpm  log2tpm Source\n",
      "285          114868_125_R_FCWC50_v1_2_RNASEQ FGFR3 15.524495 4.046534   PDMR\n",
      "295          114868_125_R_FCWC69_v1_2_RNASEQ FGFR3 27.542306 4.835030   PDMR\n",
      "305          114868_125_R_FEWC21_v1_2_RNASEQ FGFR3 17.573296 4.215158   PDMR\n",
      "315          114868_125_R_FEWC31_v1_2_RNASEQ FGFR3 29.367395 4.924451   PDMR\n",
      "605    137432_314_R_QVPG14_v2_0_1_4_0_RNASEQ FGFR3 23.924522 4.639494   PDMR\n",
      "615 137432_314_R_QVPG20HQ2_v2_0_1_4_0_RNASEQ FGFR3  2.199631 1.677905   PDMR\n",
      "    CancerGroup CancerType Formal_CaseID              NewSampleID Group\n",
      "285   Head&Neck       HNSC   PDMR-114868    PDMR-114868.P1-FCWC50   PDX\n",
      "295   Head&Neck       HNSC   PDMR-114868    PDMR-114868.P1-FCWC69   PDX\n",
      "305   Head&Neck       HNSC   PDMR-114868    PDMR-114868.P1-FEWC21   PDX\n",
      "315   Head&Neck       HNSC   PDMR-114868    PDMR-114868.P1-FEWC31   PDX\n",
      "605   Head&Neck       HNSC   PDMR-137432    PDMR-137432.P1-QVPG14   PDX\n",
      "615   Head&Neck       HNSC   PDMR-137432 PDMR-137432.P2-QVPG20HQ2   PDX\n",
      "    NCI_Passage ModelID_proper_v2   fusionName fusionType altGroup\n",
      "285          P1      114868_125_R FGFR3--TACC3    INFRAME   Fusion\n",
      "295          P1      114868_125_R FGFR3--TACC3    INFRAME   Fusion\n",
      "305          P1      114868_125_R FGFR3--TACC3    INFRAME   Fusion\n",
      "315          P1      114868_125_R FGFR3--TACC3    INFRAME   Fusion\n",
      "605          P1      137432_314_R         <NA>       <NA>       WT\n",
      "615          P2      137432_314_R         <NA>       <NA>       WT\n",
      "                                          RNASEQ gene       tpm  log2tpm Source\n",
      "472           128128_338_R_L42_v2_0_1_4_0_RNASEQ BRAF 10.928994 3.576400   PDMR\n",
      "482        128128_338_R_L43MZ1_v2_0_1_4_0_RNASEQ BRAF  5.186790 2.629191   PDMR\n",
      "492        128128_338_R_L43MZ4_v2_0_1_4_0_RNASEQ BRAF 12.098721 3.711354   PDMR\n",
      "502           128128_338_R_L43_v2_0_1_4_0_RNASEQ BRAF 14.482217 3.952540   PDMR\n",
      "512        128128_338_R_L44NT6_v2_0_1_4_0_RNASEQ BRAF 13.617304 3.869605   PDMR\n",
      "522 128128_338_R_L44NT9_RG_Q18_v2_0_1_4_0_RNASEQ BRAF  5.858963 2.777990   PDMR\n",
      "    CancerGroup CancerType Formal_CaseID              NewSampleID Group\n",
      "472        Skin       SKCM   PDMR-128128    PDMR-128128.P0-L42-v2   PDX\n",
      "482        Skin       SKCM   PDMR-128128 PDMR-128128.P1-L43MZ1-v2   PDX\n",
      "492        Skin       SKCM   PDMR-128128 PDMR-128128.P1-L43MZ4-v2   PDX\n",
      "502        Skin       SKCM   PDMR-128128    PDMR-128128.P0-L43-v2   PDX\n",
      "512        Skin       SKCM   PDMR-128128 PDMR-128128.P1-L44NT6-v2   PDX\n",
      "522        Skin       SKCM   PDMR-128128 PDMR-128128.P2-L44NT9-v2   PDX\n",
      "    NCI_Passage ModelID_proper_v2 fusionName fusionType altGroup\n",
      "472          P0      128128_338_R       <NA>       <NA>       WT\n",
      "482          P1      128128_338_R       <NA>       <NA>       WT\n",
      "492          P1      128128_338_R       <NA>       <NA>       WT\n",
      "502          P0      128128_338_R       <NA>       <NA>       WT\n",
      "512          P1      128128_338_R       <NA>       <NA>       WT\n",
      "522          P2      128128_338_R       <NA>       <NA>       WT\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in wilcox.test.default(x = 2.55252266133868, y = c(0.012350926322528, :\n",
      "“cannot compute exact p-value with ties”"
     ]
    }
   ],
   "source": [
    "d_fusion <- fread(f_fusion, data.table=F)\n",
    "\n",
    "# outdir\n",
    "outdir_sub <- \"../2.fusion_with_geneExp\"\n",
    "if (! dir.exists(outdir_sub)) { dir.create(outdir_sub) }\n",
    "\n",
    "\n",
    "#---------- prepare info\n",
    "d_sampleInfo_pdx_rna <- d_sampleInfo[d_sampleInfo$Group==\"PDX\" & d_sampleInfo$DataType==\"RNA-Seq\",]\n",
    "d_sampleInfo_pdx_rna <- d_sampleInfo_pdx_rna[,c('Analysis_ID', 'Source', 'CancerGroup', 'CancerType', 'Formal_CaseID', 'NewSampleID', 'Group', 'NCI_Passage', 'ModelID_proper_v2')]\n",
    "\n",
    "\n",
    "#---------- a) extract drug target\n",
    "fusion_tarGeneSet <- d_drugSet[d_drugSet$Event==\"Fusion\",]$Gene\n",
    "\n",
    "\n",
    "#---------- b) extract fusion for pdx\n",
    "d_fusion <- fread(f_fusion, data.table=F)\n",
    "d_fusion_pdx <- d_fusion[d_fusion$Sample %in% d_sampleInfo_pdx_rna$Analysis_ID,]\n",
    "\n",
    "\n",
    "#---------- c) format fusion data for extraction\n",
    "\n",
    "## set function for fomatting\n",
    "func_extractTargetFusion <- function(geneList, df_starFusion)\n",
    "{\n",
    "    # reset rownames\n",
    "    rownames(df_starFusion) <- NULL\n",
    "    max_row = nrow(df_starFusion)\n",
    "    \n",
    "    # new dataframe\n",
    "    d_matchedFusion <- setNames(data.frame(matrix(ncol = 4, nrow = 0)), c(\"RNASEQ\", \"gene\", \"fusionName\", \"fusionType\"))\n",
    "    \n",
    "    \n",
    "    for (i in 1:max_row){\n",
    "        \n",
    "        fusionName = df_starFusion[i,]$FusionName\n",
    "        \n",
    "        for (gene in geneList){    \n",
    "\n",
    "            if (length(grep(paste0(\"^\",gene,\"--\"), fusionName))>0 || length(grep(paste0(\"--\",gene,\"$\"), fusionName))>0){\n",
    "                RNASEQ = df_starFusion[i,]$Sample\n",
    "                fusionName = df_starFusion[i,]$FusionName\n",
    "                fusionType = df_starFusion[i,]$PROT_FUSION_TYPE\n",
    "                \n",
    "                fusion_info <- c(RNASEQ, gene, fusionName, fusionType)\n",
    "                \n",
    "                d_matchedFusion <- rbindlist(list(d_matchedFusion, as.list(fusion_info)))\n",
    "                \n",
    "            } else {\n",
    "                next\n",
    "            }\n",
    "            \n",
    "        }\n",
    "    }\n",
    "    \n",
    "    return(d_matchedFusion)\n",
    "}\n",
    "\n",
    "\n",
    "\n",
    "# extract target gene fusion of pdx samples\n",
    "d_fusion_tar <- func_extractTargetFusion(fusion_tarGeneSet, d_fusion_pdx)\n",
    "# \"RNASEQ\", \"gene\", \"fusionName\", \"fusionType\"\n",
    "\n",
    "\n",
    "# merge to one gene if the gene has multiple fusions\n",
    "library(plyr)\n",
    "d_fusion_tar <- ddply(d_fusion_tar, .(RNASEQ, gene), summarize,\n",
    "                    fusionName=paste(fusionName,collapse=\",\"), \n",
    "                    fusionType=paste(fusionType,collapse=\",\")\n",
    "                )\n",
    "# \"RNASEQ\", \"gene\", \"fusionName\", \"fusionType\"\n",
    "\n",
    "\n",
    "head(d_fusion_tar)\n",
    "\n",
    "\n",
    "#---------- d) extract gene expression\n",
    "\n",
    "# format gene exp data frame as --> sample gene log2tpm\n",
    "d_geneExp_tar <- subset(d_geneExp, Name %in% fusion_tarGeneSet)\n",
    "d_geneExp_tar <- reshape2::melt(d_geneExp_tar, id.vars=c(\"Name\"))\n",
    "colnames(d_geneExp_tar) <- c('gene', 'RNASEQ', 'tpm')\n",
    "d_geneExp_tar <- d_geneExp_tar[, c('RNASEQ', 'gene', 'tpm')]\n",
    "# RNASEQ gene tpm\n",
    "\n",
    "# tpm --> log2(tpm+1)\n",
    "d_geneExp_tar$log2tpm <- log2(d_geneExp_tar$tpm + 1)\n",
    "\n",
    "d_geneExp_tar <- d_geneExp_tar %>% filter(RNASEQ %in% d_sampleInfo_pdx_rna$Analysis_ID)\n",
    "rownames(d_geneExp_tar) <- NULL\n",
    "# RNASEQ gene tpm log2tpm\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "#---------- e) merge formatted gene exp with fusion datafrme\n",
    "\n",
    "#NOTE- must fusion data merge to gene exp rather than gene exp merge to fusion\n",
    "\n",
    "#1. merge data info\n",
    "d_tar_expFusion <- merge(d_geneExp_tar, d_sampleInfo_pdx_rna, by.x='RNASEQ', by.y=\"Analysis_ID\", all.x=TRUE)\n",
    "#2. merge fusion\n",
    "d_tar_expFusion <- merge(d_tar_expFusion, d_fusion_tar, on=c('RNASEQ', 'gene'), all.x=TRUE)\n",
    "\n",
    "\n",
    "\n",
    "#---------- f) add flag with Mut & Non-mut\n",
    "d_tar_expFusion$altGroup <- 'WT'\n",
    "d_tar_expFusion$altGroup[d_tar_expFusion$fusionName != 'NA'] <- 'Fusion'\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# output text file for checking\n",
    "write.table(d_tar_expFusion, paste0(outdir_sub,'/2.all.target.geneExp_with_fusion.',Sys.Date(),'.tsv'), sep = \"\\t\", row.names=FALSE, quote = FALSE)\n",
    "\n",
    "head(d_tar_expFusion)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "#---------- g) make visulization (1.set plot function)\n",
    "\n",
    "library(ggplot2)\n",
    "library(ggpubr)\n",
    "\n",
    "\n",
    "## two group with statistic\n",
    "CompBoxplot2exp <- function(title, dataFrame){   \n",
    "\n",
    "    #n_alt_case <- length(unique(dataFrame$Formal_CaseID[dataFrame$altGroup=='Fusion']))\n",
    "    #n_other_case <- length(unique(dataFrame$Formal_CaseID[dataFrame$altGroup=='WT']))\n",
    "    n_alt_case <- length(unique(dataFrame$ModelID_proper[dataFrame$altGroup=='Fusion']))\n",
    "    n_other_case <- length(unique(dataFrame$ModelID_proper[dataFrame$altGroup=='WT']))\n",
    "    \n",
    "    n_alt_sample <- length(unique(dataFrame$RNASEQ[dataFrame$altGroup=='Fusion']))\n",
    "    n_other_sample <- length(unique(dataFrame$RNASEQ[dataFrame$altGroup=='WT']))\n",
    "    \n",
    "    label_alt <- paste0(\"Fusion\",\"\\nN=\",n_alt_case,\"\\nn=\",n_alt_sample)\n",
    "    label_other <- paste0(\"WT\",\"\\nN=\",n_other_case,\"\\nn=\",n_other_sample)\n",
    "    \n",
    "    \n",
    "    p <- ggplot(dataFrame, aes(x=altGroup, y=log2tpm, color=altGroup)) +\n",
    "            geom_boxplot(lwd=0.6, outlier.size=0.6) + \n",
    "            geom_jitter(position=position_jitter(0.1), size=0.6, alpha = 0.5) +\n",
    "            theme_classic() +\n",
    "            scale_color_manual(values = c(\"WT\" = \"#B5B5B5\", \"Fusion\" = \"#F57E20\")) + \n",
    "            theme(legend.position=\"None\") +\n",
    "            ggtitle(title) +\n",
    "            xlab(\"\") +\n",
    "            ylab(\"Gene expression\\nlog2(TPM+1)\") + \n",
    "            scale_x_discrete(\n",
    "                limit = c(\"WT\", \"Fusion\"),\n",
    "                labels=c(\"WT\" = label_other, \"Fusion\" = label_alt)) +\n",
    "            theme(\n",
    "                plot.title = element_text(size=7, hjust = 0.5),\n",
    "                axis.text = element_text(size = 6),\n",
    "                axis.title=element_text(size=6)) + \n",
    "            theme(\n",
    "                axis.ticks = element_line(size = rel(0.5)),\n",
    "                axis.ticks.length=unit(1.5, \"mm\"),\n",
    "                axis.line = element_line(colour = 'black', size = 0.2)\n",
    "            ) +\n",
    "            theme(panel.grid.major = element_blank(), \n",
    "               panel.grid.minor = element_blank(), \n",
    "               panel.background = element_blank(), \n",
    "               panel.border = element_blank()\n",
    "            )\n",
    "\n",
    "    \n",
    "    return(p)\n",
    "\n",
    "}\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "#---------- g) make visulization (2.plotting)\n",
    "\n",
    "\n",
    "for(cancerType in unique(d_tar_expFusion$CancerType)){\n",
    "    \n",
    "    #if (cancerType!=\"SKCM\"){ next }\n",
    "    \n",
    "    df_cancer <- d_tar_expFusion[d_tar_expFusion$CancerType==cancerType,]\n",
    "    # filter sample size < 10\n",
    "    if(length(unique(df_cancer$NewSampleID))<10){ next }\n",
    "    \n",
    "    \n",
    "    pdf(paste0(outdir_sub,'/',cancerType,'.fusion_geneExp.',Sys.Date(),'.pdf'), width = 1.4, height = 1.8, useDingbats=FALSE)\n",
    "\n",
    "    for(i in unique(d_tar_expFusion$gene)){\n",
    "    \n",
    "        df2 <- df_cancer[df_cancer$gene==i,]\n",
    "        \n",
    "        # remove genes \n",
    "        # if no any mut then del. (for pdx 1 sample also possible)\n",
    "        if(nrow(df2[df2$altGroup=='Fusion',])==0){ next }\n",
    "    \n",
    "        # 2.not significant in gene exp. between two groups\n",
    "        w_test <- wilcox.test(log2tpm ~ altGroup, data = df2)\n",
    "        if(w_test$p.value >= 0.05){ next }\n",
    "    \n",
    "    \n",
    "        # output significant\n",
    "        if (nrow(df2[df2$altGroup=='Fusion',])>0 && w_test$p.value < 0.05){\n",
    "            \n",
    "            p_val <- formatC(w_test$p.value, format = \"e\", digits = 1)\n",
    "            \n",
    "           # q_val <- p.adjust(p_val, method = \"fdr\", n = length(p_val))\n",
    "            \n",
    "            title <- paste0(i, ' (', cancerType, ')', \"\\nWilcoxon, p=\", p_val)\n",
    "            \n",
    "            p1 <- CompBoxplot2exp(title, df2)\n",
    "\n",
    "            print(p1)\n",
    "            print(head(df2))\n",
    "        }\n",
    "    \n",
    "    }\n",
    "\n",
    "    dev.off()\n",
    "\n",
    "}\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<thead><tr><th scope=col>RNASEQ</th><th scope=col>gene</th><th scope=col>tpm</th><th scope=col>log2tpm</th><th scope=col>Source</th><th scope=col>CancerGroup</th><th scope=col>CancerType</th><th scope=col>Formal_CaseID</th><th scope=col>NewSampleID</th><th scope=col>Group</th><th scope=col>NCI_Passage</th><th scope=col>ModelID_proper_v2</th><th scope=col>fusionName</th><th scope=col>fusionType</th><th scope=col>altGroup</th></tr></thead>\n",
       "<tbody>\n",
       "\t<tr><td>10962X12_HCI009_PDXtumor_P3</td><td>ALK                        </td><td>  0.000000                 </td><td>0.000000                   </td><td>HCI                        </td><td>Breast                     </td><td>BRCA                       </td><td>HCI-HCI009                 </td><td>HCI-HCI009.P3              </td><td>PDX                        </td><td>P3                         </td><td>HCI009                     </td><td>NA                         </td><td>NA                         </td><td>WT                         </td></tr>\n",
       "\t<tr><td>10962X12_HCI009_PDXtumor_P3</td><td>BRAF                       </td><td>  9.491537                 </td><td>3.391154                   </td><td>HCI                        </td><td>Breast                     </td><td>BRCA                       </td><td>HCI-HCI009                 </td><td>HCI-HCI009.P3              </td><td>PDX                        </td><td>P3                         </td><td>HCI009                     </td><td>NA                         </td><td>NA                         </td><td>WT                         </td></tr>\n",
       "\t<tr><td>10962X12_HCI009_PDXtumor_P3</td><td>FGFR1                      </td><td>  7.551949                 </td><td>3.096253                   </td><td>HCI                        </td><td>Breast                     </td><td>BRCA                       </td><td>HCI-HCI009                 </td><td>HCI-HCI009.P3              </td><td>PDX                        </td><td>P3                         </td><td>HCI009                     </td><td>NA                         </td><td>NA                         </td><td>WT                         </td></tr>\n",
       "\t<tr><td>10962X12_HCI009_PDXtumor_P3</td><td>FGFR2                      </td><td>  1.076378                 </td><td>1.054069                   </td><td>HCI                        </td><td>Breast                     </td><td>BRCA                       </td><td>HCI-HCI009                 </td><td>HCI-HCI009.P3              </td><td>PDX                        </td><td>P3                         </td><td>HCI009                     </td><td>NA                         </td><td>NA                         </td><td>WT                         </td></tr>\n",
       "\t<tr><td>10962X12_HCI009_PDXtumor_P3</td><td>FGFR3                      </td><td> 26.904380                 </td><td>4.802420                   </td><td>HCI                        </td><td>Breast                     </td><td>BRCA                       </td><td>HCI-HCI009                 </td><td>HCI-HCI009.P3              </td><td>PDX                        </td><td>P3                         </td><td>HCI009                     </td><td>NA                         </td><td>NA                         </td><td>WT                         </td></tr>\n",
       "\t<tr><td>10962X12_HCI009_PDXtumor_P3</td><td>FGFR4                      </td><td>126.784226                 </td><td>6.997566                   </td><td>HCI                        </td><td>Breast                     </td><td>BRCA                       </td><td>HCI-HCI009                 </td><td>HCI-HCI009.P3              </td><td>PDX                        </td><td>P3                         </td><td>HCI009                     </td><td>NA                         </td><td>NA                         </td><td>WT                         </td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "\\begin{tabular}{r|lllllllllllllll}\n",
       " RNASEQ & gene & tpm & log2tpm & Source & CancerGroup & CancerType & Formal\\_CaseID & NewSampleID & Group & NCI\\_Passage & ModelID\\_proper\\_v2 & fusionName & fusionType & altGroup\\\\\n",
       "\\hline\n",
       "\t 10962X12\\_HCI009\\_PDXtumor\\_P3 & ALK                               &   0.000000                        & 0.000000                          & HCI                               & Breast                            & BRCA                              & HCI-HCI009                        & HCI-HCI009.P3                     & PDX                               & P3                                & HCI009                            & NA                                & NA                                & WT                               \\\\\n",
       "\t 10962X12\\_HCI009\\_PDXtumor\\_P3 & BRAF                              &   9.491537                        & 3.391154                          & HCI                               & Breast                            & BRCA                              & HCI-HCI009                        & HCI-HCI009.P3                     & PDX                               & P3                                & HCI009                            & NA                                & NA                                & WT                               \\\\\n",
       "\t 10962X12\\_HCI009\\_PDXtumor\\_P3 & FGFR1                             &   7.551949                        & 3.096253                          & HCI                               & Breast                            & BRCA                              & HCI-HCI009                        & HCI-HCI009.P3                     & PDX                               & P3                                & HCI009                            & NA                                & NA                                & WT                               \\\\\n",
       "\t 10962X12\\_HCI009\\_PDXtumor\\_P3 & FGFR2                             &   1.076378                        & 1.054069                          & HCI                               & Breast                            & BRCA                              & HCI-HCI009                        & HCI-HCI009.P3                     & PDX                               & P3                                & HCI009                            & NA                                & NA                                & WT                               \\\\\n",
       "\t 10962X12\\_HCI009\\_PDXtumor\\_P3 & FGFR3                             &  26.904380                        & 4.802420                          & HCI                               & Breast                            & BRCA                              & HCI-HCI009                        & HCI-HCI009.P3                     & PDX                               & P3                                & HCI009                            & NA                                & NA                                & WT                               \\\\\n",
       "\t 10962X12\\_HCI009\\_PDXtumor\\_P3 & FGFR4                             & 126.784226                        & 6.997566                          & HCI                               & Breast                            & BRCA                              & HCI-HCI009                        & HCI-HCI009.P3                     & PDX                               & P3                                & HCI009                            & NA                                & NA                                & WT                               \\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "RNASEQ | gene | tpm | log2tpm | Source | CancerGroup | CancerType | Formal_CaseID | NewSampleID | Group | NCI_Passage | ModelID_proper_v2 | fusionName | fusionType | altGroup | \n",
       "|---|---|---|---|---|---|\n",
       "| 10962X12_HCI009_PDXtumor_P3 | ALK                         |   0.000000                  | 0.000000                    | HCI                         | Breast                      | BRCA                        | HCI-HCI009                  | HCI-HCI009.P3               | PDX                         | P3                          | HCI009                      | NA                          | NA                          | WT                          | \n",
       "| 10962X12_HCI009_PDXtumor_P3 | BRAF                        |   9.491537                  | 3.391154                    | HCI                         | Breast                      | BRCA                        | HCI-HCI009                  | HCI-HCI009.P3               | PDX                         | P3                          | HCI009                      | NA                          | NA                          | WT                          | \n",
       "| 10962X12_HCI009_PDXtumor_P3 | FGFR1                       |   7.551949                  | 3.096253                    | HCI                         | Breast                      | BRCA                        | HCI-HCI009                  | HCI-HCI009.P3               | PDX                         | P3                          | HCI009                      | NA                          | NA                          | WT                          | \n",
       "| 10962X12_HCI009_PDXtumor_P3 | FGFR2                       |   1.076378                  | 1.054069                    | HCI                         | Breast                      | BRCA                        | HCI-HCI009                  | HCI-HCI009.P3               | PDX                         | P3                          | HCI009                      | NA                          | NA                          | WT                          | \n",
       "| 10962X12_HCI009_PDXtumor_P3 | FGFR3                       |  26.904380                  | 4.802420                    | HCI                         | Breast                      | BRCA                        | HCI-HCI009                  | HCI-HCI009.P3               | PDX                         | P3                          | HCI009                      | NA                          | NA                          | WT                          | \n",
       "| 10962X12_HCI009_PDXtumor_P3 | FGFR4                       | 126.784226                  | 6.997566                    | HCI                         | Breast                      | BRCA                        | HCI-HCI009                  | HCI-HCI009.P3               | PDX                         | P3                          | HCI009                      | NA                          | NA                          | WT                          | \n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "  RNASEQ                      gene  tpm        log2tpm  Source CancerGroup\n",
       "1 10962X12_HCI009_PDXtumor_P3 ALK     0.000000 0.000000 HCI    Breast     \n",
       "2 10962X12_HCI009_PDXtumor_P3 BRAF    9.491537 3.391154 HCI    Breast     \n",
       "3 10962X12_HCI009_PDXtumor_P3 FGFR1   7.551949 3.096253 HCI    Breast     \n",
       "4 10962X12_HCI009_PDXtumor_P3 FGFR2   1.076378 1.054069 HCI    Breast     \n",
       "5 10962X12_HCI009_PDXtumor_P3 FGFR3  26.904380 4.802420 HCI    Breast     \n",
       "6 10962X12_HCI009_PDXtumor_P3 FGFR4 126.784226 6.997566 HCI    Breast     \n",
       "  CancerType Formal_CaseID NewSampleID   Group NCI_Passage ModelID_proper_v2\n",
       "1 BRCA       HCI-HCI009    HCI-HCI009.P3 PDX   P3          HCI009           \n",
       "2 BRCA       HCI-HCI009    HCI-HCI009.P3 PDX   P3          HCI009           \n",
       "3 BRCA       HCI-HCI009    HCI-HCI009.P3 PDX   P3          HCI009           \n",
       "4 BRCA       HCI-HCI009    HCI-HCI009.P3 PDX   P3          HCI009           \n",
       "5 BRCA       HCI-HCI009    HCI-HCI009.P3 PDX   P3          HCI009           \n",
       "6 BRCA       HCI-HCI009    HCI-HCI009.P3 PDX   P3          HCI009           \n",
       "  fusionName fusionType altGroup\n",
       "1 NA         NA         WT      \n",
       "2 NA         NA         WT      \n",
       "3 NA         NA         WT      \n",
       "4 NA         NA         WT      \n",
       "5 NA         NA         WT      \n",
       "6 NA         NA         WT      "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#--------- (Important checking) make text file for record siginficant amplification\n",
    "alterationEvent <- 'Fusion'\n",
    "\n",
    "head(d_tar_expFusion)\n",
    "names(d_tar_expFusion)[names(d_tar_expFusion) == \"NewSampleID\"] <- \"SampleID\"\n",
    "\n",
    "d_rec <- func_rec_sigDiffExp_in_alt(d_tar_expFusion, alterationEvent)\n",
    "\n",
    "#print(unique(d_tar_expFusion$altGroup))\n",
    "write.table(d_rec, paste0(outdir_sub,'/1.p-val.target.geneExp_with_',alterationEvent,'.',Sys.Date(),'.tsv'), sep = \"\\t\", row.names=FALSE, quote = FALSE)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 3. CNV and gene expression (a. amplification)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<thead><tr><th scope=col>WES</th><th scope=col>gene</th><th scope=col>RNASEQ</th><th scope=col>tpm</th><th scope=col>log2tpm</th><th scope=col>Source</th><th scope=col>CancerGroup</th><th scope=col>CancerType</th><th scope=col>CaseID</th><th scope=col>SampleID</th><th scope=col>Group</th><th scope=col>Passage</th><th scope=col>segment_log2</th><th scope=col>segment_cn</th><th scope=col>altGroup</th></tr></thead>\n",
       "<tbody>\n",
       "\t<tr><td>10960X12_HCI-009_PDXtumor_P3</td><td>CCND1                       </td><td>10962X12_HCI009_PDXtumor_P3 </td><td>53.93992000                 </td><td>5.77978290                  </td><td>HCI                         </td><td>Breast                      </td><td>BRCA                        </td><td>HCI-HCI009                  </td><td>HCI-HCI009.P3               </td><td>PDX                         </td><td>P3                          </td><td>NA                          </td><td>NA                          </td><td>WT                          </td></tr>\n",
       "\t<tr><td>10960X12_HCI-009_PDXtumor_P3</td><td>CCND2                       </td><td>10962X12_HCI009_PDXtumor_P3 </td><td> 0.00795906                 </td><td>0.01143704                  </td><td>HCI                         </td><td>Breast                      </td><td>BRCA                        </td><td>HCI-HCI009                  </td><td>HCI-HCI009.P3               </td><td>PDX                         </td><td>P3                          </td><td>NA                          </td><td>NA                          </td><td>WT                          </td></tr>\n",
       "\t<tr><td>10960X12_HCI-009_PDXtumor_P3</td><td>CCND3                       </td><td>10962X12_HCI009_PDXtumor_P3 </td><td>19.95486821                 </td><td>4.38921354                  </td><td>HCI                         </td><td>Breast                      </td><td>BRCA                        </td><td>HCI-HCI009                  </td><td>HCI-HCI009.P3               </td><td>PDX                         </td><td>P3                          </td><td>NA                          </td><td>NA                          </td><td>WT                          </td></tr>\n",
       "\t<tr><td>10960X12_HCI-009_PDXtumor_P3</td><td>CDK4                        </td><td>10962X12_HCI009_PDXtumor_P3 </td><td>32.78574113                 </td><td>5.07834260                  </td><td>HCI                         </td><td>Breast                      </td><td>BRCA                        </td><td>HCI-HCI009                  </td><td>HCI-HCI009.P3               </td><td>PDX                         </td><td>P3                          </td><td>NA                          </td><td>NA                          </td><td>WT                          </td></tr>\n",
       "\t<tr><td>10960X12_HCI-009_PDXtumor_P3</td><td>CDK6                        </td><td>10962X12_HCI009_PDXtumor_P3 </td><td>11.64335100                 </td><td>3.66030698                  </td><td>HCI                         </td><td>Breast                      </td><td>BRCA                        </td><td>HCI-HCI009                  </td><td>HCI-HCI009.P3               </td><td>PDX                         </td><td>P3                          </td><td>NA                          </td><td>NA                          </td><td>WT                          </td></tr>\n",
       "\t<tr><td>10960X12_HCI-009_PDXtumor_P3</td><td>ERBB2                       </td><td>10962X12_HCI009_PDXtumor_P3 </td><td>75.91210600                 </td><td>6.26513879                  </td><td>HCI                         </td><td>Breast                      </td><td>BRCA                        </td><td>HCI-HCI009                  </td><td>HCI-HCI009.P3               </td><td>PDX                         </td><td>P3                          </td><td>NA                          </td><td>NA                          </td><td>WT                          </td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "\\begin{tabular}{r|lllllllllllllll}\n",
       " WES & gene & RNASEQ & tpm & log2tpm & Source & CancerGroup & CancerType & CaseID & SampleID & Group & Passage & segment\\_log2 & segment\\_cn & altGroup\\\\\n",
       "\\hline\n",
       "\t 10960X12\\_HCI-009\\_PDXtumor\\_P3 & CCND1                              & 10962X12\\_HCI009\\_PDXtumor\\_P3  & 53.93992000                        & 5.77978290                         & HCI                                & Breast                             & BRCA                               & HCI-HCI009                         & HCI-HCI009.P3                      & PDX                                & P3                                 & NA                                 & NA                                 & WT                                \\\\\n",
       "\t 10960X12\\_HCI-009\\_PDXtumor\\_P3 & CCND2                              & 10962X12\\_HCI009\\_PDXtumor\\_P3  &  0.00795906                        & 0.01143704                         & HCI                                & Breast                             & BRCA                               & HCI-HCI009                         & HCI-HCI009.P3                      & PDX                                & P3                                 & NA                                 & NA                                 & WT                                \\\\\n",
       "\t 10960X12\\_HCI-009\\_PDXtumor\\_P3 & CCND3                              & 10962X12\\_HCI009\\_PDXtumor\\_P3  & 19.95486821                        & 4.38921354                         & HCI                                & Breast                             & BRCA                               & HCI-HCI009                         & HCI-HCI009.P3                      & PDX                                & P3                                 & NA                                 & NA                                 & WT                                \\\\\n",
       "\t 10960X12\\_HCI-009\\_PDXtumor\\_P3 & CDK4                               & 10962X12\\_HCI009\\_PDXtumor\\_P3  & 32.78574113                        & 5.07834260                         & HCI                                & Breast                             & BRCA                               & HCI-HCI009                         & HCI-HCI009.P3                      & PDX                                & P3                                 & NA                                 & NA                                 & WT                                \\\\\n",
       "\t 10960X12\\_HCI-009\\_PDXtumor\\_P3 & CDK6                               & 10962X12\\_HCI009\\_PDXtumor\\_P3  & 11.64335100                        & 3.66030698                         & HCI                                & Breast                             & BRCA                               & HCI-HCI009                         & HCI-HCI009.P3                      & PDX                                & P3                                 & NA                                 & NA                                 & WT                                \\\\\n",
       "\t 10960X12\\_HCI-009\\_PDXtumor\\_P3 & ERBB2                              & 10962X12\\_HCI009\\_PDXtumor\\_P3  & 75.91210600                        & 6.26513879                         & HCI                                & Breast                             & BRCA                               & HCI-HCI009                         & HCI-HCI009.P3                      & PDX                                & P3                                 & NA                                 & NA                                 & WT                                \\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "WES | gene | RNASEQ | tpm | log2tpm | Source | CancerGroup | CancerType | CaseID | SampleID | Group | Passage | segment_log2 | segment_cn | altGroup | \n",
       "|---|---|---|---|---|---|\n",
       "| 10960X12_HCI-009_PDXtumor_P3 | CCND1                        | 10962X12_HCI009_PDXtumor_P3  | 53.93992000                  | 5.77978290                   | HCI                          | Breast                       | BRCA                         | HCI-HCI009                   | HCI-HCI009.P3                | PDX                          | P3                           | NA                           | NA                           | WT                           | \n",
       "| 10960X12_HCI-009_PDXtumor_P3 | CCND2                        | 10962X12_HCI009_PDXtumor_P3  |  0.00795906                  | 0.01143704                   | HCI                          | Breast                       | BRCA                         | HCI-HCI009                   | HCI-HCI009.P3                | PDX                          | P3                           | NA                           | NA                           | WT                           | \n",
       "| 10960X12_HCI-009_PDXtumor_P3 | CCND3                        | 10962X12_HCI009_PDXtumor_P3  | 19.95486821                  | 4.38921354                   | HCI                          | Breast                       | BRCA                         | HCI-HCI009                   | HCI-HCI009.P3                | PDX                          | P3                           | NA                           | NA                           | WT                           | \n",
       "| 10960X12_HCI-009_PDXtumor_P3 | CDK4                         | 10962X12_HCI009_PDXtumor_P3  | 32.78574113                  | 5.07834260                   | HCI                          | Breast                       | BRCA                         | HCI-HCI009                   | HCI-HCI009.P3                | PDX                          | P3                           | NA                           | NA                           | WT                           | \n",
       "| 10960X12_HCI-009_PDXtumor_P3 | CDK6                         | 10962X12_HCI009_PDXtumor_P3  | 11.64335100                  | 3.66030698                   | HCI                          | Breast                       | BRCA                         | HCI-HCI009                   | HCI-HCI009.P3                | PDX                          | P3                           | NA                           | NA                           | WT                           | \n",
       "| 10960X12_HCI-009_PDXtumor_P3 | ERBB2                        | 10962X12_HCI009_PDXtumor_P3  | 75.91210600                  | 6.26513879                   | HCI                          | Breast                       | BRCA                         | HCI-HCI009                   | HCI-HCI009.P3                | PDX                          | P3                           | NA                           | NA                           | WT                           | \n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "  WES                          gene  RNASEQ                      tpm        \n",
       "1 10960X12_HCI-009_PDXtumor_P3 CCND1 10962X12_HCI009_PDXtumor_P3 53.93992000\n",
       "2 10960X12_HCI-009_PDXtumor_P3 CCND2 10962X12_HCI009_PDXtumor_P3  0.00795906\n",
       "3 10960X12_HCI-009_PDXtumor_P3 CCND3 10962X12_HCI009_PDXtumor_P3 19.95486821\n",
       "4 10960X12_HCI-009_PDXtumor_P3 CDK4  10962X12_HCI009_PDXtumor_P3 32.78574113\n",
       "5 10960X12_HCI-009_PDXtumor_P3 CDK6  10962X12_HCI009_PDXtumor_P3 11.64335100\n",
       "6 10960X12_HCI-009_PDXtumor_P3 ERBB2 10962X12_HCI009_PDXtumor_P3 75.91210600\n",
       "  log2tpm    Source CancerGroup CancerType CaseID     SampleID      Group\n",
       "1 5.77978290 HCI    Breast      BRCA       HCI-HCI009 HCI-HCI009.P3 PDX  \n",
       "2 0.01143704 HCI    Breast      BRCA       HCI-HCI009 HCI-HCI009.P3 PDX  \n",
       "3 4.38921354 HCI    Breast      BRCA       HCI-HCI009 HCI-HCI009.P3 PDX  \n",
       "4 5.07834260 HCI    Breast      BRCA       HCI-HCI009 HCI-HCI009.P3 PDX  \n",
       "5 3.66030698 HCI    Breast      BRCA       HCI-HCI009 HCI-HCI009.P3 PDX  \n",
       "6 6.26513879 HCI    Breast      BRCA       HCI-HCI009 HCI-HCI009.P3 PDX  \n",
       "  Passage segment_log2 segment_cn altGroup\n",
       "1 P3      NA           NA         WT      \n",
       "2 P3      NA           NA         WT      \n",
       "3 P3      NA           NA         WT      \n",
       "4 P3      NA           NA         WT      \n",
       "5 P3      NA           NA         WT      \n",
       "6 P3      NA           NA         WT      "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "\n",
    "# outdir\n",
    "outdir_sub <- \"../3.cnv_amp_with_geneExp\"\n",
    "if (! dir.exists(outdir_sub)) { dir.create(outdir_sub) }\n",
    "\n",
    "\n",
    "#---------- a) extract drug target\n",
    "amp_tarGeneSet <- d_drugSet[d_drugSet$Event==\"Amplification\",]$Gene\n",
    "\n",
    "\n",
    "#---------- b) extract mutation\n",
    "d_cnv <- fread(f_cnv, data.table=F)\n",
    "\n",
    "# sample gene segment_log2 segment_cn gene_log2  gene_cn\n",
    "d_cnv_v2 <- d_cnv[, c('sample', 'gene', 'segment_log2', 'segment_cn')]\n",
    "\n",
    "\n",
    "d_amp_tar <- d_cnv_v2 %>%\n",
    "                filter(sample %in% d_pdx_info$WES) %>%\n",
    "                filter(gene %in% amp_tarGeneSet) %>%\n",
    "                filter(segment_cn > 5)\n",
    "\n",
    "rownames(d_amp_tar) <- NULL\n",
    "\n",
    "\n",
    "# output text file for checking\n",
    "write.table(d_amp_tar, paste0(outdir_sub,'/1a.all.target.amplification.',Sys.Date(),'.tsv'), sep = \"\\t\", row.names=FALSE, quote = FALSE)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "#---------- c) extract gene expression\n",
    "\n",
    "# format gene exp data frame as --> sample gene log2tpm\n",
    "d_geneExp_tar <- subset(d_geneExp, Name %in% amp_tarGeneSet)\n",
    "d_geneExp_tar <- reshape2::melt(d_geneExp_tar, id.vars=c(\"Name\"))\n",
    "colnames(d_geneExp_tar) <- c('gene', 'RNASEQ', 'tpm')\n",
    "d_geneExp_tar <- d_geneExp_tar[, c('RNASEQ', 'gene', 'tpm')]\n",
    "# RNASEQ gene tpm\n",
    "\n",
    "# tpm --> log2(tpm+1)\n",
    "d_geneExp_tar$log2tpm <- log2(d_geneExp_tar$tpm + 1)\n",
    "\n",
    "d_geneExp_tar <- d_geneExp_tar %>% filter(RNASEQ %in% d_pdx_info$RNASEQ)\n",
    "rownames(d_geneExp_tar) <- NULL\n",
    "# RNASEQ gene tpm log2tpm\n",
    "\n",
    "\n",
    "#---------- d) merge formatted gene exp with somatic mut datafrme\n",
    "\n",
    "#NOTE- must cnv data merge to gene exp rather than gene exp merge to cnv\n",
    "\n",
    "#1. merge data info\n",
    "d_tar_expAmp <- merge(d_geneExp_tar, d_pdx_info, on='RNASEQ', all.x=TRUE, all.y=FALSE)\n",
    "#2. merge cnv\n",
    "d_tar_expAmp <- merge(d_tar_expAmp, d_amp_tar, by.x=c('WES', 'gene'), by.y=c('sample', 'gene'), all.x=TRUE, all.y=FALSE)\n",
    "\n",
    "\n",
    "#---------- e) add flag with Mut & Non-mut\n",
    "d_tar_expAmp$altGroup <- 'WT'\n",
    "d_tar_expAmp$altGroup[d_tar_expAmp$segment_cn != 'NA'] <- 'Amplification'\n",
    "\n",
    "\n",
    "# output text file for checking\n",
    "write.table(d_tar_expAmp, paste0(outdir_sub,'/1b.all.target.geneExp_with_amplification.',Sys.Date(),'.tsv'), sep = \"\\t\", row.names=FALSE, quote = FALSE)\n",
    "\n",
    "head(d_tar_expAmp)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "#--------- (Important checking) make text file for record siginficant amplification\n",
    "alterationEvent <- 'Amplification'\n",
    "d_rec <- func_rec_sigDiffExp_in_alt(d_tar_expAmp, alterationEvent)\n",
    "write.table(d_rec, paste0(outdir_sub,'/1.p-val.target.geneExp_with_',alterationEvent,'.',Sys.Date(),'.tsv'), sep = \"\\t\", row.names=FALSE, quote = FALSE)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<strong>pdf:</strong> 2"
      ],
      "text/latex": [
       "\\textbf{pdf:} 2"
      ],
      "text/markdown": [
       "**pdf:** 2"
      ],
      "text/plain": [
       "pdf \n",
       "  2 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "\n",
    "#---------- f) make visulization (1.set plot function)\n",
    "\n",
    "library(ggplot2)\n",
    "library(ggpubr)\n",
    "\n",
    "CompBoxplot2exp_panCancers_per_gene <- function(gene, dataFrame){\n",
    "    \n",
    "    df_gene <- dataFrame[dataFrame$gene==gene,]\n",
    "    \n",
    "    max.y <- max(df_gene$log2tpm)\n",
    "    \n",
    "    p <- ggplot(df_gene, aes(x=CancerType, y=log2tpm, color=altGroup)) +\n",
    "            geom_boxplot(lwd=0.4, position=position_dodge(1), outlier.size=0.4, pch=16, outlier.alpha=0.5) + \n",
    "            geom_point(pch=16, position=position_jitterdodge(0.2), size=1, alpha = 0.5) +\n",
    "            theme_classic() +\n",
    "            #scale_color_manual(values = c(\"Amplification\" = \"#D32A29\", \"WT\" = \"#7D94AF\"), labels = c(\"Amplification\"=\"Amplification\", \"WT\"=\"No amplification\")) +\n",
    "            scale_color_manual(values = c(\"WT\" = \"#7D94AF\", \"Amplification\" = \"#D32A29\")) +\n",
    "            ggtitle(gene) +\n",
    "            theme(plot.title = element_text(size=8,hjust = 0.5),\n",
    "                    axis.title=element_text(size=7),\n",
    "                    axis.text.x = element_text(size = 7, angle = 60, hjust = 1),\n",
    "                    axis.text.y = element_text(size = 7)) + \n",
    "            theme(\n",
    "                axis.ticks = element_line(size = rel(0.5)),\n",
    "                axis.ticks.length=unit(1.5, \"mm\"),\n",
    "                axis.line = element_line(colour = 'black', size = 0.2)) +\n",
    "            xlab(\"\") +\n",
    "            ylab(\"Gene expression\\nlog2(TPM+1)\") + \n",
    "            theme(legend.position=c(0.9, 0.9),\n",
    "                    legend.title=element_blank(),\n",
    "                    legend.key.size = unit(0.5, \"cm\"),\n",
    "                    legend.text=element_text(size=7)\n",
    "            ) + \n",
    "            ylim(0,15)\n",
    "    \n",
    "    return(p)\n",
    "\n",
    "}\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "#---------- f) make visulization (2.plotting)\n",
    "\n",
    "#f_in <- paste0(outdir_sub,'/1.p-val.target.geneExp_with_Amplification','.',Sys.Date(),'.tsv')\n",
    "f_in <- '../3.cnv_amp_with_geneExp/1b.all.target.geneExp_with_amplification.filteredCase.tsv'\n",
    "\n",
    "#pdf(paste0(outdir_sub,'/','1c.all.target.geneExp_with_amplification.',Sys.Date(),'.pdf'), width = 7.5, height = 2.5, useDingbats=FALSE)\n",
    "\n",
    "pdf(paste0(f_in,'.pdf'), width = 7.5, height = 2.5, useDingbats=FALSE)\n",
    "\n",
    "\n",
    "d_tar_expAmp$altGroup <- factor(d_tar_expAmp$altGroup, levels = c('WT', 'Amplification'))\n",
    "\n",
    "for(i in unique(d_tar_expAmp$gene)){\n",
    "    \n",
    "    df2 <- d_tar_expAmp[d_tar_expAmp$gene==i,]\n",
    "    \n",
    "    if(nrow(df2[df2$altGroup=='Amplification',])==0){\n",
    "        next\n",
    "    }\n",
    "    \n",
    "    p <- CompBoxplot2exp_panCancers_per_gene(i, df2)\n",
    "    print(p)\n",
    "\n",
    "}\n",
    "\n",
    "dev.off()\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 3. CNV and gene expression (b. Deletion)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<thead><tr><th scope=col>WES</th><th scope=col>gene</th><th scope=col>RNASEQ</th><th scope=col>tpm</th><th scope=col>log2tpm</th><th scope=col>Source</th><th scope=col>CancerGroup</th><th scope=col>CancerType</th><th scope=col>CaseID</th><th scope=col>SampleID</th><th scope=col>Group</th><th scope=col>Passage</th><th scope=col>segment_log2</th><th scope=col>segment_cn</th><th scope=col>altGroup</th></tr></thead>\n",
       "<tbody>\n",
       "\t<tr><td>10960X12_HCI-009_PDXtumor_P3</td><td>MET                         </td><td>10962X12_HCI009_PDXtumor_P3 </td><td>11.29700740                 </td><td>3.62023536                  </td><td>HCI                         </td><td>Breast                      </td><td>BRCA                        </td><td>HCI-HCI009                  </td><td>HCI-HCI009.P3               </td><td>PDX                         </td><td>P3                          </td><td>NA                          </td><td>NA                          </td><td>WT                          </td></tr>\n",
       "\t<tr><td>10960X12_HCI-009_PDXtumor_P3</td><td>PTEN                        </td><td>10962X12_HCI009_PDXtumor_P3 </td><td> 0.05750863                 </td><td>0.08066944                  </td><td>HCI                         </td><td>Breast                      </td><td>BRCA                        </td><td>HCI-HCI009                  </td><td>HCI-HCI009.P3               </td><td>PDX                         </td><td>P3                          </td><td>NA                          </td><td>NA                          </td><td>WT                          </td></tr>\n",
       "\t<tr><td>10960X13_HCI-009_PDXtumor_P6</td><td>MET                         </td><td>10962X13_HCI009_PDXtumor_P6 </td><td>14.94561300                 </td><td>3.99508766                  </td><td>HCI                         </td><td>Breast                      </td><td>BRCA                        </td><td>HCI-HCI009                  </td><td>HCI-HCI009.P6               </td><td>PDX                         </td><td>P6                          </td><td>NA                          </td><td>NA                          </td><td>WT                          </td></tr>\n",
       "\t<tr><td>10960X13_HCI-009_PDXtumor_P6</td><td>PTEN                        </td><td>10962X13_HCI009_PDXtumor_P6 </td><td> 0.12914471                 </td><td>0.17523039                  </td><td>HCI                         </td><td>Breast                      </td><td>BRCA                        </td><td>HCI-HCI009                  </td><td>HCI-HCI009.P6               </td><td>PDX                         </td><td>P6                          </td><td>NA                          </td><td>NA                          </td><td>WT                          </td></tr>\n",
       "\t<tr><td>10960X14_HCI-010_PDXtumor_P3</td><td>MET                         </td><td>10962X14_HCI010_PDXtumor_P3 </td><td>30.32423900                 </td><td>4.96920756                  </td><td>HCI                         </td><td>Breast                      </td><td>BRCA                        </td><td>HCI-HCI010                  </td><td>HCI-HCI010.P3               </td><td>PDX                         </td><td>P3                          </td><td>NA                          </td><td>NA                          </td><td>WT                          </td></tr>\n",
       "\t<tr><td>10960X14_HCI-010_PDXtumor_P3</td><td>PTEN                        </td><td>10962X14_HCI010_PDXtumor_P3 </td><td> 0.58265600                 </td><td>0.66234771                  </td><td>HCI                         </td><td>Breast                      </td><td>BRCA                        </td><td>HCI-HCI010                  </td><td>HCI-HCI010.P3               </td><td>PDX                         </td><td>P3                          </td><td>NA                          </td><td>NA                          </td><td>WT                          </td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "\\begin{tabular}{r|lllllllllllllll}\n",
       " WES & gene & RNASEQ & tpm & log2tpm & Source & CancerGroup & CancerType & CaseID & SampleID & Group & Passage & segment\\_log2 & segment\\_cn & altGroup\\\\\n",
       "\\hline\n",
       "\t 10960X12\\_HCI-009\\_PDXtumor\\_P3 & MET                                & 10962X12\\_HCI009\\_PDXtumor\\_P3  & 11.29700740                        & 3.62023536                         & HCI                                & Breast                             & BRCA                               & HCI-HCI009                         & HCI-HCI009.P3                      & PDX                                & P3                                 & NA                                 & NA                                 & WT                                \\\\\n",
       "\t 10960X12\\_HCI-009\\_PDXtumor\\_P3 & PTEN                               & 10962X12\\_HCI009\\_PDXtumor\\_P3  &  0.05750863                        & 0.08066944                         & HCI                                & Breast                             & BRCA                               & HCI-HCI009                         & HCI-HCI009.P3                      & PDX                                & P3                                 & NA                                 & NA                                 & WT                                \\\\\n",
       "\t 10960X13\\_HCI-009\\_PDXtumor\\_P6 & MET                                & 10962X13\\_HCI009\\_PDXtumor\\_P6  & 14.94561300                        & 3.99508766                         & HCI                                & Breast                             & BRCA                               & HCI-HCI009                         & HCI-HCI009.P6                      & PDX                                & P6                                 & NA                                 & NA                                 & WT                                \\\\\n",
       "\t 10960X13\\_HCI-009\\_PDXtumor\\_P6 & PTEN                               & 10962X13\\_HCI009\\_PDXtumor\\_P6  &  0.12914471                        & 0.17523039                         & HCI                                & Breast                             & BRCA                               & HCI-HCI009                         & HCI-HCI009.P6                      & PDX                                & P6                                 & NA                                 & NA                                 & WT                                \\\\\n",
       "\t 10960X14\\_HCI-010\\_PDXtumor\\_P3 & MET                                & 10962X14\\_HCI010\\_PDXtumor\\_P3  & 30.32423900                        & 4.96920756                         & HCI                                & Breast                             & BRCA                               & HCI-HCI010                         & HCI-HCI010.P3                      & PDX                                & P3                                 & NA                                 & NA                                 & WT                                \\\\\n",
       "\t 10960X14\\_HCI-010\\_PDXtumor\\_P3 & PTEN                               & 10962X14\\_HCI010\\_PDXtumor\\_P3  &  0.58265600                        & 0.66234771                         & HCI                                & Breast                             & BRCA                               & HCI-HCI010                         & HCI-HCI010.P3                      & PDX                                & P3                                 & NA                                 & NA                                 & WT                                \\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "WES | gene | RNASEQ | tpm | log2tpm | Source | CancerGroup | CancerType | CaseID | SampleID | Group | Passage | segment_log2 | segment_cn | altGroup | \n",
       "|---|---|---|---|---|---|\n",
       "| 10960X12_HCI-009_PDXtumor_P3 | MET                          | 10962X12_HCI009_PDXtumor_P3  | 11.29700740                  | 3.62023536                   | HCI                          | Breast                       | BRCA                         | HCI-HCI009                   | HCI-HCI009.P3                | PDX                          | P3                           | NA                           | NA                           | WT                           | \n",
       "| 10960X12_HCI-009_PDXtumor_P3 | PTEN                         | 10962X12_HCI009_PDXtumor_P3  |  0.05750863                  | 0.08066944                   | HCI                          | Breast                       | BRCA                         | HCI-HCI009                   | HCI-HCI009.P3                | PDX                          | P3                           | NA                           | NA                           | WT                           | \n",
       "| 10960X13_HCI-009_PDXtumor_P6 | MET                          | 10962X13_HCI009_PDXtumor_P6  | 14.94561300                  | 3.99508766                   | HCI                          | Breast                       | BRCA                         | HCI-HCI009                   | HCI-HCI009.P6                | PDX                          | P6                           | NA                           | NA                           | WT                           | \n",
       "| 10960X13_HCI-009_PDXtumor_P6 | PTEN                         | 10962X13_HCI009_PDXtumor_P6  |  0.12914471                  | 0.17523039                   | HCI                          | Breast                       | BRCA                         | HCI-HCI009                   | HCI-HCI009.P6                | PDX                          | P6                           | NA                           | NA                           | WT                           | \n",
       "| 10960X14_HCI-010_PDXtumor_P3 | MET                          | 10962X14_HCI010_PDXtumor_P3  | 30.32423900                  | 4.96920756                   | HCI                          | Breast                       | BRCA                         | HCI-HCI010                   | HCI-HCI010.P3                | PDX                          | P3                           | NA                           | NA                           | WT                           | \n",
       "| 10960X14_HCI-010_PDXtumor_P3 | PTEN                         | 10962X14_HCI010_PDXtumor_P3  |  0.58265600                  | 0.66234771                   | HCI                          | Breast                       | BRCA                         | HCI-HCI010                   | HCI-HCI010.P3                | PDX                          | P3                           | NA                           | NA                           | WT                           | \n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "  WES                          gene RNASEQ                      tpm        \n",
       "1 10960X12_HCI-009_PDXtumor_P3 MET  10962X12_HCI009_PDXtumor_P3 11.29700740\n",
       "2 10960X12_HCI-009_PDXtumor_P3 PTEN 10962X12_HCI009_PDXtumor_P3  0.05750863\n",
       "3 10960X13_HCI-009_PDXtumor_P6 MET  10962X13_HCI009_PDXtumor_P6 14.94561300\n",
       "4 10960X13_HCI-009_PDXtumor_P6 PTEN 10962X13_HCI009_PDXtumor_P6  0.12914471\n",
       "5 10960X14_HCI-010_PDXtumor_P3 MET  10962X14_HCI010_PDXtumor_P3 30.32423900\n",
       "6 10960X14_HCI-010_PDXtumor_P3 PTEN 10962X14_HCI010_PDXtumor_P3  0.58265600\n",
       "  log2tpm    Source CancerGroup CancerType CaseID     SampleID      Group\n",
       "1 3.62023536 HCI    Breast      BRCA       HCI-HCI009 HCI-HCI009.P3 PDX  \n",
       "2 0.08066944 HCI    Breast      BRCA       HCI-HCI009 HCI-HCI009.P3 PDX  \n",
       "3 3.99508766 HCI    Breast      BRCA       HCI-HCI009 HCI-HCI009.P6 PDX  \n",
       "4 0.17523039 HCI    Breast      BRCA       HCI-HCI009 HCI-HCI009.P6 PDX  \n",
       "5 4.96920756 HCI    Breast      BRCA       HCI-HCI010 HCI-HCI010.P3 PDX  \n",
       "6 0.66234771 HCI    Breast      BRCA       HCI-HCI010 HCI-HCI010.P3 PDX  \n",
       "  Passage segment_log2 segment_cn altGroup\n",
       "1 P3      NA           NA         WT      \n",
       "2 P3      NA           NA         WT      \n",
       "3 P6      NA           NA         WT      \n",
       "4 P6      NA           NA         WT      \n",
       "5 P3      NA           NA         WT      \n",
       "6 P3      NA           NA         WT      "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# outdir\n",
    "outdir_sub <- \"../3.cnv_del_with_geneExp\"\n",
    "if (! dir.exists(outdir_sub)) { dir.create(outdir_sub) }\n",
    "\n",
    "altEvent <- 'Deletion'\n",
    "\n",
    "#---------- a) extract drug target\n",
    "del_tarGeneSet <- d_drugSet[d_drugSet$Event==altEvent,]$Gene\n",
    "\n",
    "\n",
    "#---------- b) extract mutation\n",
    "d_cnv <- fread(f_cnv, data.table=F)\n",
    "\n",
    "# sample gene segment_log2 segment_cn gene_log2  gene_cn\n",
    "d_cnv_v2 <- d_cnv[, c('sample', 'gene', 'segment_log2', 'segment_cn')]\n",
    "\n",
    "\n",
    "d_del_tar <- d_cnv_v2 %>%\n",
    "                filter(sample %in% d_pdx_info$WES) %>%\n",
    "                filter(gene %in% amp_tarGeneSet) %>%\n",
    "                filter(segment_log2 < -1.3)\n",
    "\n",
    "rownames(d_del_tar) <- NULL\n",
    "\n",
    "\n",
    "# output text file for checking\n",
    "write.table(d_del_tar, paste0(outdir_sub,'/2a.all.target.',altEvent,'.',Sys.Date(),'.tsv'), sep = \"\\t\", row.names=FALSE, quote = FALSE)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "#---------- c) extract gene expression\n",
    "\n",
    "# format gene exp data frame as --> sample gene log2tpm\n",
    "d_geneExp_tar <- subset(d_geneExp, Name %in% del_tarGeneSet)\n",
    "d_geneExp_tar <- reshape2::melt(d_geneExp_tar, id.vars=c(\"Name\"))\n",
    "colnames(d_geneExp_tar) <- c('gene', 'RNASEQ', 'tpm')\n",
    "d_geneExp_tar <- d_geneExp_tar[, c('RNASEQ', 'gene', 'tpm')]\n",
    "# RNASEQ gene tpm\n",
    "\n",
    "# tpm --> log2(tpm+1)\n",
    "d_geneExp_tar$log2tpm <- log2(d_geneExp_tar$tpm + 1)\n",
    "\n",
    "d_geneExp_tar <- d_geneExp_tar %>% filter(RNASEQ %in% d_pdx_info$RNASEQ)\n",
    "rownames(d_geneExp_tar) <- NULL\n",
    "# RNASEQ gene tpm log2tpm\n",
    "\n",
    "\n",
    "#---------- d) merge formatted gene exp with somatic mut datafrme\n",
    "\n",
    "#NOTE- must cnv data merge to gene exp rather than gene exp merge to cnv\n",
    "\n",
    "#1. merge data info\n",
    "d_tar_expDel <- merge(d_geneExp_tar, d_pdx_info, on='RNASEQ', all.x=TRUE, all.y=FALSE)\n",
    "#2. merge cnv\n",
    "d_tar_expDel <- merge(d_tar_expDel, d_amp_tar, by.x=c('WES', 'gene'), by.y=c('sample', 'gene'), all.x=TRUE, all.y=FALSE)\n",
    "\n",
    "\n",
    "#---------- e) add flag with Mut & Non-mut\n",
    "d_tar_expDel$altGroup <- 'WT'\n",
    "d_tar_expDel$altGroup[d_tar_expDel$segment_cn != 'NA'] <- 'Deletion'\n",
    "\n",
    "\n",
    "# output text file for checking\n",
    "write.table(d_tar_expDel, paste0(outdir_sub,'/2b.all.target.geneExp_with_', altEvent, '.',Sys.Date(),'.tsv'), sep = \"\\t\", row.names=FALSE, quote = FALSE)\n",
    "\n",
    "head(d_tar_expDel)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "#--------- (Important checking) make text file for record siginficant deletion\n",
    "alterationEvent <- 'Deletion'\n",
    "d_rec <- func_rec_sigDiffExp_in_alt(d_tar_expDel, alterationEvent)\n",
    "write.table(d_rec, paste0(outdir_sub,'/2.p-val.target.geneExp_with_',alterationEvent,'.',Sys.Date(),'.tsv'), sep = \"\\t\", row.names=FALSE, quote = FALSE)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                                WES gene                            RNASEQ\n",
      "2249     BL0382_F1232_M662_v1_2_WES  MET     BL0382_F1232_M662_v1_2_RNASEQ\n",
      "2251  BL0382_F1232_M662M18_v1_2_WES  MET  BL0382_F1232_M662M18_v1_2_RNASEQ\n",
      "2253 BL0382_F1232_M664M220_v1_2_WES  MET BL0382_F1232_M664M220_v1_2_RNASEQ\n",
      "2585    TWDE-WBC-1027-DBLSL10A_2690  MET   tblsl10a_2690.ATATCTCG-ATCTTAGT\n",
      "2587    TWDE-WBC-1027-DBLSL10B_3043  MET   tblsl10b_3043.GCGCTCTA-GCTCCGAC\n",
      "2589   TWDE-WBC-1027-DBLSL10C_3141m  MET      TWDE-WBC-1027-TBLSL10C_3141m\n",
      "2591   TWDE-WBC-1027-DBLSL10D_3280m  MET  tblsl10d_3280m.AACAGGTT-ATACCAAG\n",
      "           tpm  log2tpm Source CancerGroup CancerType      CaseID\n",
      "2249 101.30311 6.676706   PDMR     Bladder       BLCA PDMR-BL0382\n",
      "2251  73.43872 6.217981   PDMR     Bladder       BLCA PDMR-BL0382\n",
      "2253 119.39957 6.911686   PDMR     Bladder       BLCA PDMR-BL0382\n",
      "2585 430.78798 8.754179  WashU     Bladder       BLCA     WU-0029\n",
      "2587 365.02301 8.515791  WashU     Bladder       BLCA     WU-0029\n",
      "2589 428.30310 8.745853  WashU     Bladder       BLCA     WU-0029\n",
      "2591 418.97593 8.714163  WashU     Bladder       BLCA     WU-0029\n",
      "                    SampleID Group Passage segment_log2 segment_cn altGroup\n",
      "2249     PDMR-BL0382.P2-M662   PDX      P2      1.80194          7 Deletion\n",
      "2251  PDMR-BL0382.P3-M662M18   PDX      P3      1.72877          7 Deletion\n",
      "2253 PDMR-BL0382.P3-M664M220   PDX      P3      1.70868          7 Deletion\n",
      "2585              WU-0029.P0   PDX      P0      2.87578         15 Deletion\n",
      "2587              WU-0029.P1   PDX      P1      2.83560         15 Deletion\n",
      "2589              WU-0029.P2   PDX      P2      2.97526         16 Deletion\n",
      "2591              WU-0029.P3   PDX      P3      2.91683         16 Deletion\n"
     ]
    }
   ],
   "source": [
    "\n",
    "\n",
    "##======= Boxplot for two groups compared alteration with none-alteration\n",
    "\n",
    "library(ggplot2)\n",
    "library(ggpubr)\n",
    "\n",
    "\n",
    "## two group with statistic\n",
    "# dataframe\n",
    "# altGroup, log2tpm\n",
    "CompBoxplot2exp <- function(title, dataFrame){   \n",
    "    \n",
    "    n_alt_case <- length(unique(dataFrame$CaseID[dataFrame$altGroup=='Deletion']))\n",
    "    n_other_case <- length(unique(dataFrame$CaseID[dataFrame$altGroup=='WT']))\n",
    "    \n",
    "    n_alt_sample <- length(unique(dataFrame$WES[dataFrame$altGroup=='Deletion']))\n",
    "    n_other_sample <- length(unique(dataFrame$WES[dataFrame$altGroup=='WT']))\n",
    "    \n",
    "    label_alt <- paste0(\"Deletion\",\"\\nN=\",n_alt_case,\"\\nn=\",n_alt_sample)\n",
    "    label_other <- paste0(\"WT\",\"\\nN=\",n_other_case,\"\\nn=\",n_other_sample)\n",
    "    \n",
    "    \n",
    "    p <- ggplot(dataFrame, aes(x=altGroup, y=log2tpm, color=altGroup)) +\n",
    "            geom_boxplot(lwd=0.6, outlier.size=0.6) + \n",
    "            geom_jitter(position=position_jitter(0.1), size=0.6, alpha = 0.5) +\n",
    "            theme_classic() +\n",
    "            scale_color_manual(values = c(\"WT\" = \"#B5B5B5\", \"Deletion\" = \"#20255D\")) + \n",
    "            theme(legend.position=\"None\") +\n",
    "            ggtitle(title) +\n",
    "            xlab(\"\") +\n",
    "            ylab(\"Gene expression\\nlog2(TPM+1)\") + \n",
    "            scale_x_discrete(labels=c(\"Deletion\" = label_alt, \"WT\" = label_other)) +\n",
    "            theme(\n",
    "                plot.title = element_text(size=7, hjust = 0.5),\n",
    "                axis.text = element_text(size = 6),\n",
    "                axis.title=element_text(size=6)) + \n",
    "            theme(\n",
    "                axis.ticks = element_line(size = rel(0.5)),\n",
    "                axis.ticks.length=unit(1.5, \"mm\"),\n",
    "                axis.line = element_line(colour = 'black', size = 0.2)\n",
    "            ) +\n",
    "            theme(panel.grid.major = element_blank(), \n",
    "               panel.grid.minor = element_blank(), \n",
    "               panel.background = element_blank(), \n",
    "               panel.border = element_blank()\n",
    "            )\n",
    "\n",
    "    \n",
    "    return(p)\n",
    "\n",
    "}\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "#---------- f) make visulization (2.plotting)\n",
    "\n",
    "altEvent <- 'Deletion'\n",
    "\n",
    "#cancerSet <- unique(d_tar_expDel$CancerType)\n",
    "cancerSet <- c(\"BLCA\")\n",
    "\n",
    "for(cancerType in cancerSet){\n",
    "    \n",
    "    df_cancer <- d_tar_expDel[d_tar_expDel$CancerType==cancerType,]\n",
    "    # filter sample size < 10\n",
    "    if(length(unique(df_cancer$SampleID))<10){ next }\n",
    "\n",
    "    \n",
    "    pdf(paste0(outdir_sub,'/',cancerType,'.',altEvent,'_geneExp.',Sys.Date(),'.pdf'), width = 1.4, height = 1.8, useDingbats=FALSE)\n",
    "\n",
    "    for(i in unique(d_tar_expDel$gene)){\n",
    "    \n",
    "        df2 <- df_cancer[df_cancer$gene==i,]\n",
    "        \n",
    "        # remove genes \n",
    "        # if no any mut then del. (for pdx 1 sample also possible)\n",
    "        if(nrow(df2[df2$altGroup==altEvent,])==0){ next }\n",
    "    \n",
    "        # 2.not significant in gene exp. between two groups\n",
    "        w_test <- wilcox.test(log2tpm ~ altGroup, data = df2)\n",
    "        if(w_test$p.value >= 0.05){ next }\n",
    "        \n",
    "    \n",
    "        # output significant\n",
    "        if (nrow(df2[df2$altGroup==altEvent,])>0 && w_test$p.value < 0.05){\n",
    "            print(df2[df2$altGroup==altEvent,])\n",
    "            p_val <- formatC(w_test$p.value, format = \"e\", digits = 1)\n",
    "            title <- paste0(i, ' (', cancerType, ')', \"\\nWilcoxon, p=\", p_val)\n",
    "            \n",
    "            p1 <- CompBoxplot2exp(title, df2)\n",
    "\n",
    "            print(p1)\n",
    "        }\n",
    "    \n",
    "    }\n",
    "\n",
    "    dev.off()\n",
    "\n",
    "}\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<strong>pdf:</strong> 2"
      ],
      "text/latex": [
       "\\textbf{pdf:} 2"
      ],
      "text/markdown": [
       "**pdf:** 2"
      ],
      "text/plain": [
       "pdf \n",
       "  2 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "\n",
    "#---------- f) make visulization (1.set plot function)\n",
    "\n",
    "library(ggplot2)\n",
    "library(ggpubr)\n",
    "\n",
    "CompBoxplot2exp_panCancers_per_gene <- function(gene, dataFrame){\n",
    "    \n",
    "    df_gene <- dataFrame[dataFrame$gene==gene,]\n",
    "    \n",
    "    max.y <- max(df_gene$log2tpm)\n",
    "    \n",
    "    p <- ggplot(df_gene, aes(x=CancerType, y=log2tpm, color=altGroup)) +\n",
    "            geom_boxplot(lwd=0.4, position=position_dodge(1), outlier.size=0.4, pch=16, outlier.alpha=0.5) + \n",
    "            geom_point(pch=16, position=position_jitterdodge(0.2), size=1, alpha = 0.5) +\n",
    "            theme_classic() +\n",
    "            scale_color_manual(values = c( \"WT\" = \"#7D94AF\", \"Deletion\" = \"#20255D\")) + \n",
    "            ggtitle(gene) +\n",
    "            theme(plot.title = element_text(size=8,hjust = 0.5),\n",
    "                    axis.title=element_text(size=7),\n",
    "                    axis.text.x = element_text(size = 7, angle = 60, hjust = 1),\n",
    "                    axis.text.y = element_text(size = 7)) + \n",
    "            theme(\n",
    "                axis.ticks = element_line(size = rel(0.5)),\n",
    "                axis.ticks.length=unit(1.5, \"mm\"),\n",
    "                axis.line = element_line(colour = 'black', size = 0.2)) +\n",
    "            xlab(\"\") +\n",
    "            ylab(\"Gene expression\\nlog2(TPM+1)\") + \n",
    "            theme(legend.position=c(0.9, 0.9),\n",
    "                    legend.title=element_blank(),\n",
    "                    legend.key.size = unit(0.5, \"cm\"),\n",
    "                    legend.text=element_text(size=7)\n",
    "            ) + \n",
    "            ylim(0,15)\n",
    "    \n",
    "    return(p)\n",
    "\n",
    "}\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "#---------- f) make visulization (2.plotting)\n",
    "\n",
    "pdf(paste0(outdir_sub,'/','2c.all.target.geneExp_with_deltion.',Sys.Date(),'.pdf'), width = 7.5, height = 2.5, useDingbats=FALSE)\n",
    "\n",
    "for(i in unique(d_tar_expDel$gene)){\n",
    "    \n",
    "    df2 <- d_tar_expDel[d_tar_expDel$gene==i,]\n",
    "    \n",
    "    if(nrow(df2[df2$altGroup=='Deletion',])==0){\n",
    "        next\n",
    "    }\n",
    "    \n",
    "    p <- CompBoxplot2exp_panCancers_per_gene(i, df2)\n",
    "    print(p)\n",
    "\n",
    "}\n",
    "\n",
    "dev.off()\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 3. CNV and gene expression (c. Loss)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<thead><tr><th scope=col>WES</th><th scope=col>gene</th><th scope=col>RNASEQ</th><th scope=col>tpm</th><th scope=col>log2tpm</th><th scope=col>Source</th><th scope=col>CancerGroup</th><th scope=col>CancerType</th><th scope=col>CaseID</th><th scope=col>SampleID</th><th scope=col>Group</th><th scope=col>Passage</th><th scope=col>segment_log2</th><th scope=col>segment_cn</th><th scope=col>PDXmodelID</th><th scope=col>altGroup</th></tr></thead>\n",
       "<tbody>\n",
       "\t<tr><td>10960X12_HCI-009_PDXtumor_P3</td><td>MET                         </td><td>10962X12_HCI009_PDXtumor_P3 </td><td>11.29700740                 </td><td>3.62023536                  </td><td>HCI                         </td><td>Breast                      </td><td>BRCA                        </td><td>HCI-HCI009                  </td><td>HCI-HCI009.P3               </td><td>PDX                         </td><td>P3                          </td><td>       NA                   </td><td>NA                          </td><td>HCI009                      </td><td>WT                          </td></tr>\n",
       "\t<tr><td>10960X12_HCI-009_PDXtumor_P3</td><td>MLH1                        </td><td>10962X12_HCI009_PDXtumor_P3 </td><td> 8.60108803                 </td><td>3.26319791                  </td><td>HCI                         </td><td>Breast                      </td><td>BRCA                        </td><td>HCI-HCI009                  </td><td>HCI-HCI009.P3               </td><td>PDX                         </td><td>P3                          </td><td>-0.636702                   </td><td> 1                          </td><td>HCI009                      </td><td>Loss                        </td></tr>\n",
       "\t<tr><td>10960X12_HCI-009_PDXtumor_P3</td><td>MSH2                        </td><td>10962X12_HCI009_PDXtumor_P3 </td><td>11.40290670                 </td><td>3.63260636                  </td><td>HCI                         </td><td>Breast                      </td><td>BRCA                        </td><td>HCI-HCI009                  </td><td>HCI-HCI009.P3               </td><td>PDX                         </td><td>P3                          </td><td>       NA                   </td><td>NA                          </td><td>HCI009                      </td><td>WT                          </td></tr>\n",
       "\t<tr><td>10960X12_HCI-009_PDXtumor_P3</td><td>MSH6                        </td><td>10962X12_HCI009_PDXtumor_P3 </td><td>20.69513210                 </td><td>4.43929947                  </td><td>HCI                         </td><td>Breast                      </td><td>BRCA                        </td><td>HCI-HCI009                  </td><td>HCI-HCI009.P3               </td><td>PDX                         </td><td>P3                          </td><td>       NA                   </td><td>NA                          </td><td>HCI009                      </td><td>WT                          </td></tr>\n",
       "\t<tr><td>10960X12_HCI-009_PDXtumor_P3</td><td>PMS2                        </td><td>10962X12_HCI009_PDXtumor_P3 </td><td> 8.99987900                 </td><td>3.32191064                  </td><td>HCI                         </td><td>Breast                      </td><td>BRCA                        </td><td>HCI-HCI009                  </td><td>HCI-HCI009.P3               </td><td>PDX                         </td><td>P3                          </td><td>       NA                   </td><td>NA                          </td><td>HCI009                      </td><td>WT                          </td></tr>\n",
       "\t<tr><td>10960X12_HCI-009_PDXtumor_P3</td><td>PTEN                        </td><td>10962X12_HCI009_PDXtumor_P3 </td><td> 0.05750863                 </td><td>0.08066944                  </td><td>HCI                         </td><td>Breast                      </td><td>BRCA                        </td><td>HCI-HCI009                  </td><td>HCI-HCI009.P3               </td><td>PDX                         </td><td>P3                          </td><td>-6.803390                   </td><td> 0                          </td><td>HCI009                      </td><td>Loss                        </td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "\\begin{tabular}{r|llllllllllllllll}\n",
       " WES & gene & RNASEQ & tpm & log2tpm & Source & CancerGroup & CancerType & CaseID & SampleID & Group & Passage & segment\\_log2 & segment\\_cn & PDXmodelID & altGroup\\\\\n",
       "\\hline\n",
       "\t 10960X12\\_HCI-009\\_PDXtumor\\_P3 & MET                                & 10962X12\\_HCI009\\_PDXtumor\\_P3  & 11.29700740                        & 3.62023536                         & HCI                                & Breast                             & BRCA                               & HCI-HCI009                         & HCI-HCI009.P3                      & PDX                                & P3                                 &        NA                          & NA                                 & HCI009                             & WT                                \\\\\n",
       "\t 10960X12\\_HCI-009\\_PDXtumor\\_P3 & MLH1                               & 10962X12\\_HCI009\\_PDXtumor\\_P3  &  8.60108803                        & 3.26319791                         & HCI                                & Breast                             & BRCA                               & HCI-HCI009                         & HCI-HCI009.P3                      & PDX                                & P3                                 & -0.636702                          &  1                                 & HCI009                             & Loss                              \\\\\n",
       "\t 10960X12\\_HCI-009\\_PDXtumor\\_P3 & MSH2                               & 10962X12\\_HCI009\\_PDXtumor\\_P3  & 11.40290670                        & 3.63260636                         & HCI                                & Breast                             & BRCA                               & HCI-HCI009                         & HCI-HCI009.P3                      & PDX                                & P3                                 &        NA                          & NA                                 & HCI009                             & WT                                \\\\\n",
       "\t 10960X12\\_HCI-009\\_PDXtumor\\_P3 & MSH6                               & 10962X12\\_HCI009\\_PDXtumor\\_P3  & 20.69513210                        & 4.43929947                         & HCI                                & Breast                             & BRCA                               & HCI-HCI009                         & HCI-HCI009.P3                      & PDX                                & P3                                 &        NA                          & NA                                 & HCI009                             & WT                                \\\\\n",
       "\t 10960X12\\_HCI-009\\_PDXtumor\\_P3 & PMS2                               & 10962X12\\_HCI009\\_PDXtumor\\_P3  &  8.99987900                        & 3.32191064                         & HCI                                & Breast                             & BRCA                               & HCI-HCI009                         & HCI-HCI009.P3                      & PDX                                & P3                                 &        NA                          & NA                                 & HCI009                             & WT                                \\\\\n",
       "\t 10960X12\\_HCI-009\\_PDXtumor\\_P3 & PTEN                               & 10962X12\\_HCI009\\_PDXtumor\\_P3  &  0.05750863                        & 0.08066944                         & HCI                                & Breast                             & BRCA                               & HCI-HCI009                         & HCI-HCI009.P3                      & PDX                                & P3                                 & -6.803390                          &  0                                 & HCI009                             & Loss                              \\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "WES | gene | RNASEQ | tpm | log2tpm | Source | CancerGroup | CancerType | CaseID | SampleID | Group | Passage | segment_log2 | segment_cn | PDXmodelID | altGroup | \n",
       "|---|---|---|---|---|---|\n",
       "| 10960X12_HCI-009_PDXtumor_P3 | MET                          | 10962X12_HCI009_PDXtumor_P3  | 11.29700740                  | 3.62023536                   | HCI                          | Breast                       | BRCA                         | HCI-HCI009                   | HCI-HCI009.P3                | PDX                          | P3                           |        NA                    | NA                           | HCI009                       | WT                           | \n",
       "| 10960X12_HCI-009_PDXtumor_P3 | MLH1                         | 10962X12_HCI009_PDXtumor_P3  |  8.60108803                  | 3.26319791                   | HCI                          | Breast                       | BRCA                         | HCI-HCI009                   | HCI-HCI009.P3                | PDX                          | P3                           | -0.636702                    |  1                           | HCI009                       | Loss                         | \n",
       "| 10960X12_HCI-009_PDXtumor_P3 | MSH2                         | 10962X12_HCI009_PDXtumor_P3  | 11.40290670                  | 3.63260636                   | HCI                          | Breast                       | BRCA                         | HCI-HCI009                   | HCI-HCI009.P3                | PDX                          | P3                           |        NA                    | NA                           | HCI009                       | WT                           | \n",
       "| 10960X12_HCI-009_PDXtumor_P3 | MSH6                         | 10962X12_HCI009_PDXtumor_P3  | 20.69513210                  | 4.43929947                   | HCI                          | Breast                       | BRCA                         | HCI-HCI009                   | HCI-HCI009.P3                | PDX                          | P3                           |        NA                    | NA                           | HCI009                       | WT                           | \n",
       "| 10960X12_HCI-009_PDXtumor_P3 | PMS2                         | 10962X12_HCI009_PDXtumor_P3  |  8.99987900                  | 3.32191064                   | HCI                          | Breast                       | BRCA                         | HCI-HCI009                   | HCI-HCI009.P3                | PDX                          | P3                           |        NA                    | NA                           | HCI009                       | WT                           | \n",
       "| 10960X12_HCI-009_PDXtumor_P3 | PTEN                         | 10962X12_HCI009_PDXtumor_P3  |  0.05750863                  | 0.08066944                   | HCI                          | Breast                       | BRCA                         | HCI-HCI009                   | HCI-HCI009.P3                | PDX                          | P3                           | -6.803390                    |  0                           | HCI009                       | Loss                         | \n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "  WES                          gene RNASEQ                      tpm        \n",
       "1 10960X12_HCI-009_PDXtumor_P3 MET  10962X12_HCI009_PDXtumor_P3 11.29700740\n",
       "2 10960X12_HCI-009_PDXtumor_P3 MLH1 10962X12_HCI009_PDXtumor_P3  8.60108803\n",
       "3 10960X12_HCI-009_PDXtumor_P3 MSH2 10962X12_HCI009_PDXtumor_P3 11.40290670\n",
       "4 10960X12_HCI-009_PDXtumor_P3 MSH6 10962X12_HCI009_PDXtumor_P3 20.69513210\n",
       "5 10960X12_HCI-009_PDXtumor_P3 PMS2 10962X12_HCI009_PDXtumor_P3  8.99987900\n",
       "6 10960X12_HCI-009_PDXtumor_P3 PTEN 10962X12_HCI009_PDXtumor_P3  0.05750863\n",
       "  log2tpm    Source CancerGroup CancerType CaseID     SampleID      Group\n",
       "1 3.62023536 HCI    Breast      BRCA       HCI-HCI009 HCI-HCI009.P3 PDX  \n",
       "2 3.26319791 HCI    Breast      BRCA       HCI-HCI009 HCI-HCI009.P3 PDX  \n",
       "3 3.63260636 HCI    Breast      BRCA       HCI-HCI009 HCI-HCI009.P3 PDX  \n",
       "4 4.43929947 HCI    Breast      BRCA       HCI-HCI009 HCI-HCI009.P3 PDX  \n",
       "5 3.32191064 HCI    Breast      BRCA       HCI-HCI009 HCI-HCI009.P3 PDX  \n",
       "6 0.08066944 HCI    Breast      BRCA       HCI-HCI009 HCI-HCI009.P3 PDX  \n",
       "  Passage segment_log2 segment_cn PDXmodelID altGroup\n",
       "1 P3             NA    NA         HCI009     WT      \n",
       "2 P3      -0.636702     1         HCI009     Loss    \n",
       "3 P3             NA    NA         HCI009     WT      \n",
       "4 P3             NA    NA         HCI009     WT      \n",
       "5 P3             NA    NA         HCI009     WT      \n",
       "6 P3      -6.803390     0         HCI009     Loss    "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "\n",
    "# outdir\n",
    "outdir_sub <- \"../3.cnv_loss_with_geneExp\"\n",
    "if (! dir.exists(outdir_sub)) { dir.create(outdir_sub) }\n",
    "\n",
    "\n",
    "#---------- a) extract drug target\n",
    "#loss_tarGeneSet <- d_drugSet[d_drugSet$Event==altEvent,]$Gene\n",
    "loss_tarGeneSet <- d_drugSet[d_drugSet$Event %in% c('Loss','Deletion'),]$Gene\n",
    "\n",
    "\n",
    "#---------- b) extract mutation\n",
    "d_cnv <- fread(f_cnv, data.table=F)\n",
    "\n",
    "# sample gene segment_log2 segment_cn gene_log2  gene_cn\n",
    "d_cnv_v2 <- d_cnv[, c('sample', 'gene', 'segment_log2', 'segment_cn')]\n",
    "\n",
    "\n",
    "d_loss_tar <- d_cnv_v2 %>%\n",
    "                filter(sample %in% d_pdx_info$WES) %>%\n",
    "                filter(gene %in% loss_tarGeneSet) %>%\n",
    "                filter(segment_log2 < -0.4)\n",
    "                #filter(segment_cn ==1)\n",
    "\n",
    "rownames(d_loss_tar) <- NULL\n",
    "\n",
    "\n",
    "# output text file for checking\n",
    "#write.table(d_loss_tar, paste0(outdir_sub,'/3a.all.target.loss.',Sys.Date(),'.tsv'), sep = \"\\t\", row.names=FALSE, quote = FALSE)\n",
    "write.table(d_loss_tar, paste0(outdir_sub,'/3a.all.target.loss_del.',Sys.Date(),'.tsv'), sep = \"\\t\", row.names=FALSE, quote = FALSE)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "#---------- c) extract gene expression\n",
    "\n",
    "# format gene exp data frame as --> sample gene log2tpm\n",
    "d_geneExp_tar <- subset(d_geneExp, Name %in% loss_tarGeneSet)\n",
    "d_geneExp_tar <- reshape2::melt(d_geneExp_tar, id.vars=c(\"Name\"))\n",
    "colnames(d_geneExp_tar) <- c('gene', 'RNASEQ', 'tpm')\n",
    "d_geneExp_tar <- d_geneExp_tar[, c('RNASEQ', 'gene', 'tpm')]\n",
    "# RNASEQ gene tpm\n",
    "\n",
    "# tpm --> log2(tpm+1)\n",
    "d_geneExp_tar$log2tpm <- log2(d_geneExp_tar$tpm + 1)\n",
    "\n",
    "d_geneExp_tar <- d_geneExp_tar %>% filter(RNASEQ %in% d_pdx_info$RNASEQ)\n",
    "rownames(d_geneExp_tar) <- NULL\n",
    "# RNASEQ gene tpm log2tpm\n",
    "\n",
    "\n",
    "#---------- d) merge formatted gene exp with somatic mut datafrme\n",
    "\n",
    "#NOTE- must cnv data merge to gene exp rather than gene exp merge to cnv\n",
    "\n",
    "#1. merge data info\n",
    "d_tar_expLoss <- merge(d_geneExp_tar, d_pdx_info, on='RNASEQ', all.x=TRUE, all.y=FALSE)\n",
    "#2. merge cnv\n",
    "d_tar_expLoss <- merge(d_tar_expLoss, d_loss_tar, by.x=c('WES', 'gene'), by.y=c('sample', 'gene'), all.x=TRUE, all.y=FALSE)\n",
    "\n",
    "# 3. add PDX model ID\n",
    "d_tar_expLoss$PDXmodelID <- d_sampleInfo$ModelID_proper[match(d_tar_expLoss$WES, d_sampleInfo$Analysis_ID)]\n",
    "\n",
    "#---------- e) add flag with Mut & Non-mut\n",
    "d_tar_expLoss$altGroup <- 'WT'\n",
    "d_tar_expLoss$altGroup[d_tar_expLoss$segment_cn != 'NA'] <- 'Loss'\n",
    "\n",
    "\n",
    "# output text file for checking\n",
    "#write.table(d_tar_expLoss, paste0(outdir_sub,'/3b.all.target.geneExp_with_loss.',Sys.Date(),'.tsv'), sep = \"\\t\", row.names=FALSE, quote = FALSE)\n",
    "write.table(d_tar_expLoss, paste0(outdir_sub,'/3b.all.target.geneExp_with_loss_del.',Sys.Date(),'.tsv'), sep = \"\\t\", row.names=FALSE, quote = FALSE)\n",
    "\n",
    "\n",
    "\n",
    "head(d_tar_expLoss)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in wilcox.test.default(x = c(4.0909638710484, 4.41350846046682, :\n",
      "“cannot compute exact p-value with ties”Warning message in wilcox.test.default(x = c(7.09538218326582, 2.60070459010433, :\n",
      "“cannot compute exact p-value with ties”Warning message in wilcox.test.default(x = c(0.11535379202118, 0.0998152221225018, :\n",
      "“cannot compute exact p-value with ties”Warning message in wilcox.test.default(x = c(0, 0, 0, 6.39309239868405, 5.74573308057332, :\n",
      "“cannot compute exact p-value with ties”"
     ]
    }
   ],
   "source": [
    "#--------- (Important checking) make text file for record siginficant deletion\n",
    "alterationEvent <- 'Loss'\n",
    "d_rec <- func_rec_sigDiffExp_in_alt(d_tar_expLoss, alterationEvent)\n",
    "write.table(d_rec, paste0(outdir_sub,'/3.p-val.target.geneExp_with_',alterationEvent,'.',Sys.Date(),'.tsv'), sep = \"\\t\", row.names=FALSE, quote = FALSE)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<strong>pdf:</strong> 2"
      ],
      "text/latex": [
       "\\textbf{pdf:} 2"
      ],
      "text/markdown": [
       "**pdf:** 2"
      ],
      "text/plain": [
       "pdf \n",
       "  2 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "\n",
    "#---------- f) make visulization (1.set plot function)\n",
    "\n",
    "library(ggplot2)\n",
    "library(ggpubr)\n",
    "\n",
    "CompBoxplot2exp_panCancers_per_gene <- function(gene, dataFrame){\n",
    "    \n",
    "    df_gene <- dataFrame[dataFrame$gene==gene,]\n",
    "    \n",
    "    max.y <- max(df_gene$log2tpm)\n",
    "    \n",
    "    p <- ggplot(df_gene, aes(x=CancerType, y=log2tpm, color=altGroup)) +\n",
    "            geom_boxplot(lwd=0.4, position=position_dodge(1), outlier.size=0.4, pch=16, outlier.alpha=0.5) + \n",
    "            geom_point(pch=16, position=position_jitterdodge(0.2), size=1, alpha = 0.5) +\n",
    "            theme_classic() +\n",
    "            scale_color_manual(values = c(\"WT\" = \"#7D94AF\", \"Loss\" = \"#98509F\")) + \n",
    "            ggtitle(gene) +\n",
    "            theme(plot.title = element_text(size=8,hjust = 0.5),\n",
    "                    axis.title=element_text(size=7),\n",
    "                    axis.text.x = element_text(size = 7, angle = 60, hjust = 1),\n",
    "                    axis.text.y = element_text(size = 7)) + \n",
    "            theme(\n",
    "                axis.ticks = element_line(size = rel(0.5)),\n",
    "                axis.ticks.length=unit(1.5, \"mm\"),\n",
    "                axis.line = element_line(colour = 'black', size = 0.2)) +\n",
    "            xlab(\"\") +\n",
    "            ylab(\"Gene expression\\nlog2(TPM+1)\") + \n",
    "            theme(legend.position=c(0.9, 0.9),\n",
    "                    legend.title=element_blank(),\n",
    "                    legend.key.size = unit(0.5, \"cm\"),\n",
    "                    legend.text=element_text(size=7)\n",
    "            ) + \n",
    "            theme(panel.grid.major = element_blank(), \n",
    "               panel.grid.minor = element_blank(), \n",
    "               panel.background = element_blank(), \n",
    "               panel.border = element_blank()\n",
    "            ) +\n",
    "            ylim(0,15)\n",
    "    \n",
    "    return(p)\n",
    "\n",
    "}\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "#---------- f) make visulization (2.plotting)\n",
    "\n",
    "#pdf(paste0(outdir_sub,'/','3c.all.target.geneExp_with_loss.',Sys.Date(),'.pdf'), width = 7.5, height = 2.5, useDingbats=FALSE)\n",
    "pdf(paste0(outdir_sub,'/','3c.all.target.geneExp_with_loss_del.',Sys.Date(),'.pdf'), width = 7.5, height = 2.5, useDingbats=FALSE)\n",
    "\n",
    "\n",
    "for(i in unique(d_tar_expLoss$gene)){\n",
    "    \n",
    "    df2 <- d_tar_expLoss[d_tar_expLoss$gene==i,]\n",
    "    \n",
    "    if(nrow(df2[df2$altGroup=='Loss',])==0){\n",
    "        next\n",
    "    }\n",
    "    \n",
    "    p <- CompBoxplot2exp_panCancers_per_gene(i, df2)\n",
    "    print(p)\n",
    "\n",
    "}\n",
    "\n",
    "dev.off()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<thead><tr><th scope=col>WES</th><th scope=col>gene</th><th scope=col>RNASEQ</th><th scope=col>tpm</th><th scope=col>log2tpm</th><th scope=col>Source</th><th scope=col>CancerGroup</th><th scope=col>CancerType</th><th scope=col>CaseID</th><th scope=col>SampleID</th><th scope=col>Group</th><th scope=col>Passage</th><th scope=col>segment_log2</th><th scope=col>segment_cn</th><th scope=col>PDXmodelID</th><th scope=col>altGroup</th></tr></thead>\n",
       "<tbody>\n",
       "\t<tr><td>10960X12_HCI-009_PDXtumor_P3</td><td>MET                         </td><td>10962X12_HCI009_PDXtumor_P3 </td><td>11.29700740                 </td><td>3.62023536                  </td><td>HCI                         </td><td>Breast                      </td><td>BRCA                        </td><td>HCI-HCI009                  </td><td>HCI-HCI009.P3               </td><td>PDX                         </td><td>P3                          </td><td>       NA                   </td><td>NA                          </td><td>HCI009                      </td><td>WT                          </td></tr>\n",
       "\t<tr><td>10960X12_HCI-009_PDXtumor_P3</td><td>MLH1                        </td><td>10962X12_HCI009_PDXtumor_P3 </td><td> 8.60108803                 </td><td>3.26319791                  </td><td>HCI                         </td><td>Breast                      </td><td>BRCA                        </td><td>HCI-HCI009                  </td><td>HCI-HCI009.P3               </td><td>PDX                         </td><td>P3                          </td><td>-0.636702                   </td><td> 1                          </td><td>HCI009                      </td><td>Loss                        </td></tr>\n",
       "\t<tr><td>10960X12_HCI-009_PDXtumor_P3</td><td>MSH2                        </td><td>10962X12_HCI009_PDXtumor_P3 </td><td>11.40290670                 </td><td>3.63260636                  </td><td>HCI                         </td><td>Breast                      </td><td>BRCA                        </td><td>HCI-HCI009                  </td><td>HCI-HCI009.P3               </td><td>PDX                         </td><td>P3                          </td><td>       NA                   </td><td>NA                          </td><td>HCI009                      </td><td>WT                          </td></tr>\n",
       "\t<tr><td>10960X12_HCI-009_PDXtumor_P3</td><td>MSH6                        </td><td>10962X12_HCI009_PDXtumor_P3 </td><td>20.69513210                 </td><td>4.43929947                  </td><td>HCI                         </td><td>Breast                      </td><td>BRCA                        </td><td>HCI-HCI009                  </td><td>HCI-HCI009.P3               </td><td>PDX                         </td><td>P3                          </td><td>       NA                   </td><td>NA                          </td><td>HCI009                      </td><td>WT                          </td></tr>\n",
       "\t<tr><td>10960X12_HCI-009_PDXtumor_P3</td><td>PMS2                        </td><td>10962X12_HCI009_PDXtumor_P3 </td><td> 8.99987900                 </td><td>3.32191064                  </td><td>HCI                         </td><td>Breast                      </td><td>BRCA                        </td><td>HCI-HCI009                  </td><td>HCI-HCI009.P3               </td><td>PDX                         </td><td>P3                          </td><td>       NA                   </td><td>NA                          </td><td>HCI009                      </td><td>WT                          </td></tr>\n",
       "\t<tr><td>10960X12_HCI-009_PDXtumor_P3</td><td>PTEN                        </td><td>10962X12_HCI009_PDXtumor_P3 </td><td> 0.05750863                 </td><td>0.08066944                  </td><td>HCI                         </td><td>Breast                      </td><td>BRCA                        </td><td>HCI-HCI009                  </td><td>HCI-HCI009.P3               </td><td>PDX                         </td><td>P3                          </td><td>-6.803390                   </td><td> 0                          </td><td>HCI009                      </td><td>Loss                        </td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "\\begin{tabular}{r|llllllllllllllll}\n",
       " WES & gene & RNASEQ & tpm & log2tpm & Source & CancerGroup & CancerType & CaseID & SampleID & Group & Passage & segment\\_log2 & segment\\_cn & PDXmodelID & altGroup\\\\\n",
       "\\hline\n",
       "\t 10960X12\\_HCI-009\\_PDXtumor\\_P3 & MET                                & 10962X12\\_HCI009\\_PDXtumor\\_P3  & 11.29700740                        & 3.62023536                         & HCI                                & Breast                             & BRCA                               & HCI-HCI009                         & HCI-HCI009.P3                      & PDX                                & P3                                 &        NA                          & NA                                 & HCI009                             & WT                                \\\\\n",
       "\t 10960X12\\_HCI-009\\_PDXtumor\\_P3 & MLH1                               & 10962X12\\_HCI009\\_PDXtumor\\_P3  &  8.60108803                        & 3.26319791                         & HCI                                & Breast                             & BRCA                               & HCI-HCI009                         & HCI-HCI009.P3                      & PDX                                & P3                                 & -0.636702                          &  1                                 & HCI009                             & Loss                              \\\\\n",
       "\t 10960X12\\_HCI-009\\_PDXtumor\\_P3 & MSH2                               & 10962X12\\_HCI009\\_PDXtumor\\_P3  & 11.40290670                        & 3.63260636                         & HCI                                & Breast                             & BRCA                               & HCI-HCI009                         & HCI-HCI009.P3                      & PDX                                & P3                                 &        NA                          & NA                                 & HCI009                             & WT                                \\\\\n",
       "\t 10960X12\\_HCI-009\\_PDXtumor\\_P3 & MSH6                               & 10962X12\\_HCI009\\_PDXtumor\\_P3  & 20.69513210                        & 4.43929947                         & HCI                                & Breast                             & BRCA                               & HCI-HCI009                         & HCI-HCI009.P3                      & PDX                                & P3                                 &        NA                          & NA                                 & HCI009                             & WT                                \\\\\n",
       "\t 10960X12\\_HCI-009\\_PDXtumor\\_P3 & PMS2                               & 10962X12\\_HCI009\\_PDXtumor\\_P3  &  8.99987900                        & 3.32191064                         & HCI                                & Breast                             & BRCA                               & HCI-HCI009                         & HCI-HCI009.P3                      & PDX                                & P3                                 &        NA                          & NA                                 & HCI009                             & WT                                \\\\\n",
       "\t 10960X12\\_HCI-009\\_PDXtumor\\_P3 & PTEN                               & 10962X12\\_HCI009\\_PDXtumor\\_P3  &  0.05750863                        & 0.08066944                         & HCI                                & Breast                             & BRCA                               & HCI-HCI009                         & HCI-HCI009.P3                      & PDX                                & P3                                 & -6.803390                          &  0                                 & HCI009                             & Loss                              \\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "WES | gene | RNASEQ | tpm | log2tpm | Source | CancerGroup | CancerType | CaseID | SampleID | Group | Passage | segment_log2 | segment_cn | PDXmodelID | altGroup | \n",
       "|---|---|---|---|---|---|\n",
       "| 10960X12_HCI-009_PDXtumor_P3 | MET                          | 10962X12_HCI009_PDXtumor_P3  | 11.29700740                  | 3.62023536                   | HCI                          | Breast                       | BRCA                         | HCI-HCI009                   | HCI-HCI009.P3                | PDX                          | P3                           |        NA                    | NA                           | HCI009                       | WT                           | \n",
       "| 10960X12_HCI-009_PDXtumor_P3 | MLH1                         | 10962X12_HCI009_PDXtumor_P3  |  8.60108803                  | 3.26319791                   | HCI                          | Breast                       | BRCA                         | HCI-HCI009                   | HCI-HCI009.P3                | PDX                          | P3                           | -0.636702                    |  1                           | HCI009                       | Loss                         | \n",
       "| 10960X12_HCI-009_PDXtumor_P3 | MSH2                         | 10962X12_HCI009_PDXtumor_P3  | 11.40290670                  | 3.63260636                   | HCI                          | Breast                       | BRCA                         | HCI-HCI009                   | HCI-HCI009.P3                | PDX                          | P3                           |        NA                    | NA                           | HCI009                       | WT                           | \n",
       "| 10960X12_HCI-009_PDXtumor_P3 | MSH6                         | 10962X12_HCI009_PDXtumor_P3  | 20.69513210                  | 4.43929947                   | HCI                          | Breast                       | BRCA                         | HCI-HCI009                   | HCI-HCI009.P3                | PDX                          | P3                           |        NA                    | NA                           | HCI009                       | WT                           | \n",
       "| 10960X12_HCI-009_PDXtumor_P3 | PMS2                         | 10962X12_HCI009_PDXtumor_P3  |  8.99987900                  | 3.32191064                   | HCI                          | Breast                       | BRCA                         | HCI-HCI009                   | HCI-HCI009.P3                | PDX                          | P3                           |        NA                    | NA                           | HCI009                       | WT                           | \n",
       "| 10960X12_HCI-009_PDXtumor_P3 | PTEN                         | 10962X12_HCI009_PDXtumor_P3  |  0.05750863                  | 0.08066944                   | HCI                          | Breast                       | BRCA                         | HCI-HCI009                   | HCI-HCI009.P3                | PDX                          | P3                           | -6.803390                    |  0                           | HCI009                       | Loss                         | \n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "  WES                          gene RNASEQ                      tpm        \n",
       "1 10960X12_HCI-009_PDXtumor_P3 MET  10962X12_HCI009_PDXtumor_P3 11.29700740\n",
       "2 10960X12_HCI-009_PDXtumor_P3 MLH1 10962X12_HCI009_PDXtumor_P3  8.60108803\n",
       "3 10960X12_HCI-009_PDXtumor_P3 MSH2 10962X12_HCI009_PDXtumor_P3 11.40290670\n",
       "4 10960X12_HCI-009_PDXtumor_P3 MSH6 10962X12_HCI009_PDXtumor_P3 20.69513210\n",
       "5 10960X12_HCI-009_PDXtumor_P3 PMS2 10962X12_HCI009_PDXtumor_P3  8.99987900\n",
       "6 10960X12_HCI-009_PDXtumor_P3 PTEN 10962X12_HCI009_PDXtumor_P3  0.05750863\n",
       "  log2tpm    Source CancerGroup CancerType CaseID     SampleID      Group\n",
       "1 3.62023536 HCI    Breast      BRCA       HCI-HCI009 HCI-HCI009.P3 PDX  \n",
       "2 3.26319791 HCI    Breast      BRCA       HCI-HCI009 HCI-HCI009.P3 PDX  \n",
       "3 3.63260636 HCI    Breast      BRCA       HCI-HCI009 HCI-HCI009.P3 PDX  \n",
       "4 4.43929947 HCI    Breast      BRCA       HCI-HCI009 HCI-HCI009.P3 PDX  \n",
       "5 3.32191064 HCI    Breast      BRCA       HCI-HCI009 HCI-HCI009.P3 PDX  \n",
       "6 0.08066944 HCI    Breast      BRCA       HCI-HCI009 HCI-HCI009.P3 PDX  \n",
       "  Passage segment_log2 segment_cn PDXmodelID altGroup\n",
       "1 P3             NA    NA         HCI009     WT      \n",
       "2 P3      -0.636702     1         HCI009     Loss    \n",
       "3 P3             NA    NA         HCI009     WT      \n",
       "4 P3             NA    NA         HCI009     WT      \n",
       "5 P3             NA    NA         HCI009     WT      \n",
       "6 P3      -6.803390     0         HCI009     Loss    "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "\n",
    "##=================== Make loss vs no loss boxplot for specific gene and cancerType\n",
    "\n",
    "\n",
    "# outdir\n",
    "outdir_sub <- \"../3.cnv_loss_with_geneExp\"\n",
    "if (! dir.exists(outdir_sub)) { dir.create(outdir_sub) }\n",
    "\n",
    "\n",
    "\n",
    "#---------- g) make visulization (1.set plot function)\n",
    "\n",
    "library(ggplot2)\n",
    "library(ggpubr)\n",
    "\n",
    "\n",
    "## two group with statistic\n",
    "CompBoxplot2exp <- function(title, dataFrame){   \n",
    "    \n",
    "    #n_alt_case <- length(unique(dataFrame$CaseID[dataFrame$altGroup=='Loss']))\n",
    "    #n_other_case <- length(unique(dataFrame$CaseID[dataFrame$altGroup=='WT']))\n",
    "    n_alt_case <- length(unique(dataFrame$PDXmodelID[dataFrame$altGroup=='Loss']))\n",
    "    n_other_case <- length(unique(dataFrame$PDXmodelID[dataFrame$altGroup=='WT']))\n",
    "    \n",
    "    n_alt_sample <- length(unique(dataFrame$WES[dataFrame$altGroup=='Loss']))\n",
    "    n_other_sample <- length(unique(dataFrame$WES[dataFrame$altGroup=='WT']))\n",
    "    \n",
    "    label_alt <- paste0(\"Loss\",\"\\nN=\",n_alt_case,\"\\nn=\",n_alt_sample)\n",
    "    label_other <- paste0(\"WT\",\"\\nN=\",n_other_case,\"\\nn=\",n_other_sample)\n",
    "    \n",
    "    \n",
    "    p <- ggplot(dataFrame, aes(x=altGroup, y=log2tpm, color=altGroup)) +\n",
    "            geom_boxplot(lwd=0.6, outlier.size=0.6) + \n",
    "            geom_jitter(position=position_jitter(0.1), size=0.6, alpha = 0.5) +\n",
    "            theme_classic() +\n",
    "            scale_color_manual(values = c(\"WT\" = \"#B5B5B5\", \"Loss\" = \"#98509F\")) + \n",
    "            theme(legend.position=\"None\") +\n",
    "            ggtitle(title) +\n",
    "            xlab(\"\") +\n",
    "            ylab(\"Gene expression\\nlog2(TPM+1)\") + \n",
    "            scale_x_discrete(\n",
    "                limit = c(\"WT\", \"Loss\"),\n",
    "                labels=c(\"Loss\" = label_alt, \"WT\" = label_other)) +\n",
    "            theme(\n",
    "                plot.title = element_text(size=7, hjust = 0.5),\n",
    "                axis.text = element_text(size = 6),\n",
    "                axis.title=element_text(size=6)) + \n",
    "            theme(\n",
    "                axis.ticks = element_line(size = rel(0.5)),\n",
    "                axis.ticks.length=unit(1.5, \"mm\"),\n",
    "                axis.line = element_line(colour = 'black', size = 0.2)\n",
    "            ) +\n",
    "            theme(panel.grid.major = element_blank(), \n",
    "               panel.grid.minor = element_blank(), \n",
    "               panel.background = element_blank(), \n",
    "               panel.border = element_blank()\n",
    "            )\n",
    "\n",
    "    \n",
    "    return(p)\n",
    "\n",
    "}\n",
    "\n",
    "\n",
    "\n",
    "#---------- g) make visulization (2.plotting)\n",
    "\n",
    "cancerTypeSet <- c('SARC', 'SKCM', 'PAAD', 'COAD')\n",
    "\n",
    "for(cancerType in cancerTypeSet){\n",
    "    \n",
    "    df_cancer <- d_tar_expLoss[d_tar_expLoss$CancerType==cancerType,]\n",
    "    # filter sample size < 10\n",
    "    if(length(unique(df_cancer$SampleID))<10){ next }\n",
    "    \n",
    "    \n",
    "    pdf(paste0(outdir_sub,'/',cancerType,'.loss_geneExp.',Sys.Date(),'.pdf'), width = 1.4, height = 1.8, useDingbats=FALSE)\n",
    "\n",
    "    for(i in unique(d_tar_expLoss$gene)){\n",
    "    \n",
    "        df2 <- df_cancer[df_cancer$gene==i,]\n",
    "        \n",
    "        # remove genes \n",
    "        # if no any mut then del. (for pdx 1 sample also possible)\n",
    "        if(nrow(df2[df2$altGroup=='Loss',])==0){ next }\n",
    "    \n",
    "        # 2.not significant in gene exp. between two groups\n",
    "        w_test <- wilcox.test(log2tpm ~ altGroup, data = df2)\n",
    "        if(w_test$p.value >= 0.01){ next }\n",
    "    \n",
    "    \n",
    "        # output significant\n",
    "        if (nrow(df2[df2$altGroup=='Loss',])>0 && w_test$p.value < 0.01){\n",
    "            \n",
    "            p_val <- formatC(w_test$p.value, format = \"e\", digits = 1)\n",
    "            title <- paste0(i, ' (', cancerType, ')', \"\\nWilcoxon, p=\", p_val)\n",
    "            \n",
    "            p1 <- CompBoxplot2exp(title, df2)\n",
    "\n",
    "            print(p1)\n",
    "        }\n",
    "    \n",
    "    }\n",
    "\n",
    "    dev.off()\n",
    "\n",
    "}\n",
    "\n",
    "head(d_tar_expLoss)\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "3.5.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
