

## CharGer

CharGer (Characterization of Germline variants) is a software tool for interpreting and predicting clinical pathogenicity of germline variants. CharGer gathers evidence from databases and annotations, provided by local tools and files or via ReST APIs, and classifies variants according to ACMG guidelines for assessing variant pathogenicity. User-designed pathogenicity criteria can be incorporated into CharGer’s flexible framework, thereby allowing users to create a customized classification protocol.

https://github.com/ding-lab/CharGer/tree/v0.5.4


## Full CharGer documentation

https://ding-lab.github.io/CharGer/contents.html#mastertoc


## Cross-reference data files used in PDX PanCan Germline analyses:

We implemented CharGer using the parameters as indicated in the run_CharGer_on_VEP.sh script.
The cross-reference data files used here are the same as used by Huang et al (2018), in order to obtain better cancer-relevant results (https://doi.org/10.1016/j.cell.2018.03.039)
The cross-reference data files used are described below. They can be found in the CharGer v0.5.4 github page (https://github.com/ding-lab/CharGer/tree/v0.5.4/PanCanAtlasData), or under the directory `PanCanAtlasData` in this page.

#### Files and associated parameters:

`-z pathogenic variants`

  - **file:** `PanCanAtlasData/emptyRemoved_20160428_pathogenic_variants_HGVSg_VEP_grch38lifOver.vcf`

  - **description:** compiled list of known pathogenic variants from Huang et al (2018). Variants that result in identical peptide changes as variants known pathogenic in ClinVar or this compiled list are assigned a PS1 score. Variants at the same position but that result in a different peptide change are assigned a PM5 score.
  
 `--inheritanceGeneList inheritance gene list file, (format: gene\tdisease\tmode_of_inheritance)`

  - **file:** `PanCanAtlasData/20160301_Rahman_KJ_KH_gene_table_CharGer.txt`

  - **description:** list of 152 cancer predisposition genes compiled by Huang et al (2018) and their mode of inheritance. Variants in these genes are assigned to different evidence levels based on variant type and mode of inheritance: 
    - truncations in genes carrying variants with a dominant mode of inheritance are assigned PVS1; recessive variants in these genes are assigned PSC1. We only assigned this evidence to truncations in tumor suppressor genes but not oncogenes. 
    - protein length changes due to inframe insertions or deletions or nonstop variants in genes carrying variants with a dominant mode of inheritance receive a PM4, whereas recessives receive a PPC1. 
 
 `--PP2GeneList PP2 gene list file, (format: column of genes) .txt`

  - **file:** `PanCanAtlasData/152_cpgs.txt`

  - **description:**  list of 152 cancer predisposition genes compiled by Huang et al (2018). Missense variants in these genes are assigned to PP2 evidence level.
  

`-H HotSpot3D clusters file, .clusters`
  
  - **file:** PanCanAtlasData/MC3.noHypers.mericUnspecified.d10.r20.v114.grch38liftOver.clusters

  - **description:** file generated using our HotSpot3D tool (Niu et al., 2016); this file was generated by Huang et al (2018) using the TCGA cohort and was also used here. Briefly, HotSpot3D was run on TCGA MC3 somatic mutation calls (hypermutators removed), identifying mutation clusters enriched by recurrent and neighboring pockets of mutations. Germline variants found to be a somatic mutation with recurrence in at least two samples among all cancer types in a HotSpot3D cluster were assigned to PM1 evidence level.

## References:

_CharGer:_
  Scott, A. D., Huang, K. L., Weerasinghe, A., Mashl, R. J., Gao, Q., Martins Rodrigues, F., ... & Ding, L. (2019). CharGer: clinical Characterization of Germline variants. Bioinformatics, 35(5), 865-867.

_PanCancer TCGA Germline Study (Huang et al, 2018):_ 
  Huang, K. L., Mashl, R. J., Wu, Y., Ritter, D. I., Wang, J., Oh, C., Paczkowska, M., Reynolds, S., Wyczalkowski, M. A., Oak, N., Scott, A. D., Krassowski, M., Cherniack, A. D., Houlahan, K. E., Jayasinghe, R., Wang, L. B., Zhou, D. C., Liu, D., Cao, S., Kim, Y. W., … Ding, L. (2018). **Pathogenic Germline Variants in 10,389 Adult Cancers**. Cell, 173(2), 355–370.e14. https://doi.org/10.1016/j.cell.2018.03.039

_HotSpot3D:_
  Niu, B., Scott, A. D., Sengupta, S., Bailey, M. H., Batra, P., Ning, J., ... & Ding, L. (2016). **Protein-structure-guided discovery of functional mutations across 19 cancer types.** Nature genetics, 48(8), 827-837.


Data analysis: Martins Rodrigues, Fernanda <fernanda@wustl.edu>  and  Terekhanova, Nadezhda <nvterekhanova@wustl.edu>


